{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### In this Jupyter Note Book we will read Files which are converted in 0_Reduced_indels from Original Intervar file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging Case Control SNP Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30\n",
      "30\n"
     ]
    }
   ],
   "source": [
    "case = ['1119','910','873','1489','989','880','1282','1584','875','1009','MWB_110',\n",
    "        'MWB_161','MWB_174','MWB_242','MWB_276','MWB_437','MWB_490','MWB_506',\n",
    "        'MWB_540','MWB_548','MWB_564','MWB_626','MWB_719','MWB_740','MWB_746',\n",
    "        'MWB_815','MWB_870','MWB_1226','MWB_1276',\n",
    "        'MWB_1445']\n",
    "\n",
    "\n",
    "control = ['1093','1134','1293','1316','1389','1428','1477','1481','1557','1591','MWB_895','MWB_911',\n",
    "           'MWB_941','MWB_956','MWB_961',\"MWB_1008\",'MWB_1040','MWB_1151','MWB_1244',\n",
    "           'MWB_1250','MWB_1443','MWB_1676','MWB_1960','MWB_1964','MWB_1975',\n",
    "           'MWB_1977','MWB_1982','MWB_1995','MWB_2007','MWB_2008']\n",
    "\n",
    "print(len(case))\n",
    "print(len(control))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def split(x):\n",
    "    \"This Function will return only Front 1 or 2 word\"\n",
    "    splitted = x.rsplit()\n",
    "    if splitted[0] in [\"Pathogenic\", \"Benign\"]:\n",
    "        return splitted[:1][0]\n",
    "    else:\n",
    "        return \"_\".join(splitted[:2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### This will save the File, _indels.csv and _snv.csv "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping 1093-->Skipping 1134-->Skipping 1293-->Skipping 1316-->Skipping 1389-->Skipping 1428-->Skipping 1477-->Skipping 1481-->Skipping 1557-->Skipping 1591-->Skipping MWB_895-->Skipping MWB_911-->Skipping MWB_941-->Skipping MWB_956-->Skipping MWB_961-->Skipping MWB_1008-->Skipping MWB_1040-->Skipping MWB_1151-->Skipping MWB_1244-->Skipping MWB_1250-->Skipping MWB_1443-->Skipping MWB_1676-->Skipping MWB_1960-->Skipping MWB_1964-->Skipping MWB_1975-->Skipping MWB_1977-->Skipping MWB_1982-->Skipping MWB_1995-->Skipping MWB_2007-->Skipping MWB_2008-->"
     ]
    }
   ],
   "source": [
    "for _ in control:\n",
    "    # If file exist will skip\n",
    "    condition = os.path.exists(_+\"/\"+_+\"_indels.csv\")\n",
    "    if not condition:\n",
    "        print(\"Doing\", _)\n",
    "        df = pd.read_csv(_+\"/\"+_+\"_snv.csv\", low_memory = False)\n",
    "        df[\"InterVar\"] = df.InterVar.apply(lambda x : split(x))\n",
    "        df.to_csv(_+\"/\"+_+\"_snv.csv\", index = False)\n",
    "    \n",
    "        df = pd.read_csv(_+\"/\"+_+\"_indels.csv\", low_memory = False)\n",
    "        df[\"InterVar\"] = df.InterVar.apply(lambda x : split(x))\n",
    "        df.to_csv(_+\"/\"+_+\"_indels.csv\", index = False)\n",
    "        del(df)\n",
    "    else:\n",
    "        print(\"Skipping\", _, end= \"-->\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping 1119-->Skipping 910-->Skipping 873-->Skipping 1489-->Skipping 989-->Skipping 880-->Skipping 1282-->Skipping 1584-->Skipping 875-->Skipping 1009-->Skipping MWB_110-->Skipping MWB_161-->Skipping MWB_174-->Skipping MWB_242-->Skipping MWB_276-->Skipping MWB_437-->Skipping MWB_490-->Skipping MWB_506-->Skipping MWB_540-->Skipping MWB_548-->Skipping MWB_564-->Skipping MWB_626-->Skipping MWB_719-->Skipping MWB_740-->Skipping MWB_746-->Skipping MWB_815-->Skipping MWB_870-->Skipping MWB_1226-->Skipping MWB_1276-->Skipping MWB_1445-->"
     ]
    }
   ],
   "source": [
    "# Same Process For case\n",
    "\n",
    "for _ in case:\n",
    "    # If file exist will skip\n",
    "    condition = os.path.exists(_+\"/\"+_+\"_indels.csv\")\n",
    "    if not condition:\n",
    "        print(\"Doing\", _)\n",
    "        df = pd.read_csv(_+\"/\"+_+\"_snv.csv\", low_memory = False)\n",
    "        df[\"InterVar\"] = df.InterVar.apply(lambda x : split(x))\n",
    "        df.to_csv(_+\"/\"+_+\"_snv.csv\", index = False)\n",
    "    \n",
    "        df = pd.read_csv(_+\"/\"+_+\"_indels.csv\", low_memory = False)\n",
    "        df[\"InterVar\"] = df.InterVar.apply(lambda x : split(x))\n",
    "        df.to_csv(_+\"/\"+_+\"_indels.csv\", index = False)\n",
    "        del(df)\n",
    "    else:\n",
    "        print(\"Skipping\", _, end= \"-->\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#case_merged = None\n",
    "#open_cols = [\"Chr\", 'Start', 'Ref',\"Alt\"]\n",
    "#merged_cols = [\"Chr\", 'Start', 'Ref',\"Alt\",\"SNP\"]\n",
    "#case_merged = pd.read_csv(\"1119/1119_snv.csv\", low_memory= False,usecols=open_cols, dtype=\"object\")\n",
    "#case_merged.drop_duplicates(subset = [\"Start\",\"Alt\"], keep = False, inplace = True)\n",
    "#case_merged[\"SNP\"] = 1\n",
    "#case_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read all Files and Merged as a single File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done in the begining\n",
      "Skipping 910\n",
      "Skipping 873\n",
      "Skipping 1489\n",
      "Skipping 989\n",
      "Skipping 880\n",
      "Skipping 1282\n",
      "Skipping 1584\n",
      "Skipping 875\n",
      "Skipping 1009\n",
      "Skipping MWB_110\n",
      "Skipping MWB_161\n",
      "Skipping MWB_174\n",
      "Skipping MWB_242\n",
      "Skipping MWB_276\n",
      "Skipping MWB_437\n",
      "Skipping MWB_490\n",
      "Skipping MWB_506\n",
      "Skipping MWB_540\n",
      "Skipping MWB_548\n",
      "Skipping MWB_564\n",
      "Skipping MWB_626\n",
      "Skipping MWB_719\n",
      "Skipping MWB_740\n",
      "Skipping MWB_746\n",
      "Skipping MWB_815\n",
      "Skipping MWB_870\n",
      "Skipping MWB_1226\n",
      "Skipping MWB_1276\n",
      "Skipping MWB_1445\n"
     ]
    }
   ],
   "source": [
    "for _ in case:\n",
    "    if _ == \"1119\":\n",
    "        print(\"Done in the begining\")\n",
    "        continue\n",
    "    # If file exist will skip\n",
    "    condition = os.path.exists(_+\"/\"+_+\"_indels.csv\")\n",
    "    \n",
    "    if not condition:    \n",
    "        print(\"DOinfg\",_)\n",
    "        df = pd.read_csv(_+\"/\"+_+\"_snv.csv\", usecols = open_cols, low_memory=False, dtype=\"object\")\n",
    "        df.drop_duplicates(subset = \"Start\", keep = False, inplace = True)\n",
    "    \n",
    "        df[\"SNP_\"] = 1\n",
    "        case_merged = pd.merge(case_merged, df, on= open_cols, how = \"outer\")\n",
    "        case_merged.fillna(0, inplace = True)\n",
    "    \n",
    "        case_merged[\"SNP\"] = case_merged.SNP + case_merged.SNP_\n",
    "        #case_merged.drop_duplicates(subset = \"Start\", keep = False, inplace = True)\n",
    "        case_merged.drop([\"SNP_\"], axis = 1, inplace=True)\n",
    "        del df\n",
    "    else:\n",
    "        print(\"Skipping\", _)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "case_merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#case_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "case_merged.drop_duplicates(subset = [\"Start\",\"Alt\"], keep = False, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#case_merged.to_csv(\"30_case_mearged_snv.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#case_merged[case_merged.SNP >29]\n",
    "\n",
    "merged_cols = [\"Chr\", 'Start', 'Ref',\"Alt\",\"SNP_Case\"]\n",
    "case_merged.columns = merged_cols\n",
    "\n",
    "merged_cols = [\"Chr\", 'Start', 'Ref',\"Alt\",\"SNP_Control\"]\n",
    "control_merged.columns = merged_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Since File already Saved so lets Open and test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\krish\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3063: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "# Load 30_control_merged and merge 2 File as a single File\n",
    "control_merged = pd.read_csv(\"30_control_mearged_snv.csv\")\n",
    "case_merged = pd.read_csv(\"30_case_mearged_snv.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chr       object\n",
       "Start      int64\n",
       "Ref       object\n",
       "Alt       object\n",
       "SNP      float64\n",
       "dtype: object"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case_merged.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chr       object\n",
       "Start      int64\n",
       "Ref       object\n",
       "Alt       object\n",
       "SNP      float64\n",
       "dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_merged.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>SNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14907</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14930</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>15211</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>15274</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>15274</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chr  Start Ref Alt   SNP\n",
       "0   1  14907   A   G  18.0\n",
       "1   1  14930   A   G  18.0\n",
       "2   1  15211   T   G  24.0\n",
       "3   1  15274   A   G  26.0\n",
       "4   1  15274   A   T  24.0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>SNP_Control</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14907</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14930</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>15211</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>15274</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>15274</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chr  Start Ref Alt  SNP_Control\n",
       "0   1  14907   A   G         18.0\n",
       "1   1  14930   A   G         18.0\n",
       "2   1  15211   T   G         24.0\n",
       "3   1  15274   A   G         26.0\n",
       "4   1  15274   A   T         24.0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_cols = [\"Chr\", 'Start', 'Ref',\"Alt\",\"SNP_Control\"]\n",
    "control_merged.columns = merged_cols\n",
    "control_merged.head()\n",
    "#case_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>SNP_Case</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14907</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14930</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>15118</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>15211</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>15274</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4391656</th>\n",
       "      <td>X</td>\n",
       "      <td>154762862</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4391657</th>\n",
       "      <td>X</td>\n",
       "      <td>154764201</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4391658</th>\n",
       "      <td>X</td>\n",
       "      <td>154851612</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4391659</th>\n",
       "      <td>Y</td>\n",
       "      <td>5308865</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4391660</th>\n",
       "      <td>Y</td>\n",
       "      <td>5458247</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4391661 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr      Start Ref Alt  SNP_Case\n",
       "0         1      14907   A   G      13.0\n",
       "1         1      14930   A   G      17.0\n",
       "2         1      15118   A   G      15.0\n",
       "3         1      15211   T   G      25.0\n",
       "4         1      15274   A   G       5.0\n",
       "...      ..        ...  ..  ..       ...\n",
       "4391656   X  154762862   C   T       1.0\n",
       "4391657   X  154764201   T   G       1.0\n",
       "4391658   X  154851612   C   A       1.0\n",
       "4391659   Y    5308865   G   A       1.0\n",
       "4391660   Y    5458247   G   A       1.0\n",
       "\n",
       "[4391661 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_cols = [\"Chr\", 'Start', 'Ref',\"Alt\",\"SNP_Case\"]\n",
    "case_merged.columns = merged_cols\n",
    "case_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#control_merged.Start = control_merged.Start.astype(object)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "open_cols = [\"Chr\", 'Start', 'Ref',\"Alt\"]\n",
    "_merged = pd.merge(case_merged, control_merged, on= open_cols, how = \"outer\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Mistakely there are  \"-\"  in ALT column so lets drop these\n",
    "\n",
    "_merged.drop(_merged[_merged.Alt == \"-\"].index, axis = 0, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>SNP_Case</th>\n",
       "      <th>SNP_Control</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14907</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>13.0</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14930</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>17.0</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>15118</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>15.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>15211</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>25.0</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>15274</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>5.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chr  Start Ref Alt  SNP_Case  SNP_Control\n",
       "0   1  14907   A   G      13.0         18.0\n",
       "1   1  14930   A   G      17.0         18.0\n",
       "2   1  15118   A   G      15.0          NaN\n",
       "3   1  15211   T   G      25.0         24.0\n",
       "4   1  15274   A   G       5.0         26.0"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#_merged.fillna(0, inplace= True)\n",
    "_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#_merged.to_csv(\"case_control_merged_Intervar.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>SNP_Case</th>\n",
       "      <th>SNP_Control</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>125810</th>\n",
       "      <td>1</td>\n",
       "      <td>246412158</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143688</th>\n",
       "      <td>2</td>\n",
       "      <td>33016361</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143689</th>\n",
       "      <td>2</td>\n",
       "      <td>33016474</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143691</th>\n",
       "      <td>2</td>\n",
       "      <td>33016812</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320055</th>\n",
       "      <td>3</td>\n",
       "      <td>128887527</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329876</th>\n",
       "      <td>3</td>\n",
       "      <td>149288035</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377525</th>\n",
       "      <td>4</td>\n",
       "      <td>38524051</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384065</th>\n",
       "      <td>4</td>\n",
       "      <td>54260937</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384104</th>\n",
       "      <td>4</td>\n",
       "      <td>54305404</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384124</th>\n",
       "      <td>4</td>\n",
       "      <td>54321613</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414195</th>\n",
       "      <td>4</td>\n",
       "      <td>115532601</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>24.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>489957</th>\n",
       "      <td>5</td>\n",
       "      <td>98119864</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>489958</th>\n",
       "      <td>5</td>\n",
       "      <td>98125530</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>25.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632729</th>\n",
       "      <td>7</td>\n",
       "      <td>13982060</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>25.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950753</th>\n",
       "      <td>11</td>\n",
       "      <td>17902999</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950754</th>\n",
       "      <td>11</td>\n",
       "      <td>17904634</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950755</th>\n",
       "      <td>11</td>\n",
       "      <td>17905278</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950756</th>\n",
       "      <td>11</td>\n",
       "      <td>17905323</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950759</th>\n",
       "      <td>11</td>\n",
       "      <td>17906294</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950770</th>\n",
       "      <td>11</td>\n",
       "      <td>17912080</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950771</th>\n",
       "      <td>11</td>\n",
       "      <td>17912156</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950774</th>\n",
       "      <td>11</td>\n",
       "      <td>17917897</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950779</th>\n",
       "      <td>11</td>\n",
       "      <td>17923860</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1077466</th>\n",
       "      <td>12</td>\n",
       "      <td>118515520</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>24.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1079770</th>\n",
       "      <td>12</td>\n",
       "      <td>122881563</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>24.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1090199</th>\n",
       "      <td>13</td>\n",
       "      <td>21336486</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103304</th>\n",
       "      <td>13</td>\n",
       "      <td>39980168</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>25.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1213630</th>\n",
       "      <td>15</td>\n",
       "      <td>66365220</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1278735</th>\n",
       "      <td>16</td>\n",
       "      <td>81565749</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1278739</th>\n",
       "      <td>16</td>\n",
       "      <td>81566448</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1278741</th>\n",
       "      <td>16</td>\n",
       "      <td>81566640</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1278743</th>\n",
       "      <td>16</td>\n",
       "      <td>81567115</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1278745</th>\n",
       "      <td>16</td>\n",
       "      <td>81567245</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1278746</th>\n",
       "      <td>16</td>\n",
       "      <td>81567554</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1510025</th>\n",
       "      <td>X</td>\n",
       "      <td>100226530</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1557723</th>\n",
       "      <td>1</td>\n",
       "      <td>231358301</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1557727</th>\n",
       "      <td>1</td>\n",
       "      <td>231360962</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1557730</th>\n",
       "      <td>1</td>\n",
       "      <td>231362923</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1557731</th>\n",
       "      <td>1</td>\n",
       "      <td>231363983</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1557734</th>\n",
       "      <td>1</td>\n",
       "      <td>231367940</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1557736</th>\n",
       "      <td>1</td>\n",
       "      <td>231368406</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1821182</th>\n",
       "      <td>10</td>\n",
       "      <td>27692561</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1821184</th>\n",
       "      <td>10</td>\n",
       "      <td>27693868</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>24.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1821185</th>\n",
       "      <td>10</td>\n",
       "      <td>27694654</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>24.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1821187</th>\n",
       "      <td>10</td>\n",
       "      <td>27698535</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>24.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1821189</th>\n",
       "      <td>10</td>\n",
       "      <td>27700424</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>24.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1821190</th>\n",
       "      <td>10</td>\n",
       "      <td>27700615</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>24.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1821192</th>\n",
       "      <td>10</td>\n",
       "      <td>27702624</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>24.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1821206</th>\n",
       "      <td>10</td>\n",
       "      <td>27807199</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>24.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1821216</th>\n",
       "      <td>10</td>\n",
       "      <td>27826474</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>25.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1887048</th>\n",
       "      <td>12</td>\n",
       "      <td>122962286</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>25.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2151696</th>\n",
       "      <td>6</td>\n",
       "      <td>57396040</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr      Start Ref Alt  SNP_Case  SNP_Control\n",
       "125810    1  246412158   T   C      24.0         12.0\n",
       "143688    2   33016361   G   C      24.0         12.0\n",
       "143689    2   33016474   G   C      24.0         12.0\n",
       "143691    2   33016812   T   C      24.0         12.0\n",
       "320055    3  128887527   A   G      24.0         12.0\n",
       "329876    3  149288035   G   A      24.0         12.0\n",
       "377525    4   38524051   A   G      24.0         12.0\n",
       "384065    4   54260937   C   A      24.0         12.0\n",
       "384104    4   54305404   A   G      24.0         12.0\n",
       "384124    4   54321613   G   A      24.0         12.0\n",
       "414195    4  115532601   T   C      24.0         11.0\n",
       "489957    5   98119864   A   G      24.0         12.0\n",
       "489958    5   98125530   T   G      25.0         12.0\n",
       "632729    7   13982060   A   G      25.0         12.0\n",
       "950753   11   17902999   C   G      24.0         12.0\n",
       "950754   11   17904634   A   G      24.0         12.0\n",
       "950755   11   17905278   C   G      24.0         12.0\n",
       "950756   11   17905323   A   G      24.0         12.0\n",
       "950759   11   17906294   T   A      24.0         12.0\n",
       "950770   11   17912080   G   C      24.0         12.0\n",
       "950771   11   17912156   T   C      24.0         12.0\n",
       "950774   11   17917897   G   A      24.0         12.0\n",
       "950779   11   17923860   T   G      24.0         12.0\n",
       "1077466  12  118515520   C   T      24.0         10.0\n",
       "1079770  12  122881563   C   T      24.0         11.0\n",
       "1090199  13   21336486   C   T      24.0         12.0\n",
       "1103304  13   39980168   G   A      25.0         12.0\n",
       "1213630  15   66365220   G   A      24.0         12.0\n",
       "1278735  16   81565749   A   G      24.0         11.0\n",
       "1278739  16   81566448   T   C      24.0         12.0\n",
       "1278741  16   81566640   G   C      24.0         12.0\n",
       "1278743  16   81567115   T   G      24.0         12.0\n",
       "1278745  16   81567245   A   G      24.0         12.0\n",
       "1278746  16   81567554   A   G      24.0         12.0\n",
       "1510025   X  100226530   G   A      24.0         12.0\n",
       "1557723   1  231358301   G   A      24.0         12.0\n",
       "1557727   1  231360962   T   C      24.0         12.0\n",
       "1557730   1  231362923   G   A      24.0         12.0\n",
       "1557731   1  231363983   T   G      24.0         12.0\n",
       "1557734   1  231367940   G   A      24.0         12.0\n",
       "1557736   1  231368406   A   G      24.0         12.0\n",
       "1821182  10   27692561   T   G      24.0         11.0\n",
       "1821184  10   27693868   C   T      24.0         11.0\n",
       "1821185  10   27694654   C   T      24.0         11.0\n",
       "1821187  10   27698535   T   C      24.0         11.0\n",
       "1821189  10   27700424   T   C      24.0         11.0\n",
       "1821190  10   27700615   C   T      24.0         11.0\n",
       "1821192  10   27702624   G   T      24.0         11.0\n",
       "1821206  10   27807199   G   T      24.0         12.0\n",
       "1821216  10   27826474   A   G      25.0         12.0\n",
       "1887048  12  122962286   G   A      25.0         12.0\n",
       "2151696   6   57396040   A   G      24.0         11.0"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_merged[(_merged.SNP_Case >23) & (_merged.SNP_Control <=12)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lets do IN Control Again"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "case_merged = None  # To clear Memory\n",
    "control_merged = None\n",
    "open_cols = [\"Chr\", 'Start', 'Ref',\"Alt\"]\n",
    "merged_cols = [\"Chr\", 'Start', 'Ref',\"Alt\",\"SNP\"]\n",
    "control_merged = pd.read_csv(\"1093/1093_snv.csv\", low_memory= False,usecols=open_cols, dtype=\"object\")\n",
    "control_merged.drop_duplicates(subset =  [\"Start\",\"Alt\"], keep = False, inplace = True)\n",
    "control_merged[\"SNP\"] = 1\n",
    "print(control_merged.shape)\n",
    "control_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _ in control:\n",
    "    if _ == \"1093\":\n",
    "        print(\"Done in the begining\")\n",
    "        continue\n",
    "    print(_)\n",
    "    df = pd.read_csv(_+\"/\"+_+\"_snv.csv\", usecols = open_cols, low_memory=False, dtype=\"object\")\n",
    "    df.drop_duplicates(subset = [\"Start\",\"Alt\"], keep = False, inplace = True)\n",
    "    \n",
    "    df[\"SNP_\"] = 1\n",
    "    control_merged = pd.merge(control_merged, df, on= open_cols, how = \"outer\")\n",
    "    control_merged.fillna(0, inplace = True)\n",
    "    \n",
    "    control_merged[\"SNP\"] = control_merged.SNP + control_merged.SNP_\n",
    "    #case_merged.drop_duplicates(subset = \"Start\", keep = False, inplace = True)\n",
    "    control_merged.drop([\"SNP_\"], axis = 1, inplace=True)\n",
    "    del df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#control_merged.to_csv(\"30_control_mearged_snv.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "control_merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = ['Chr', 'Start', 'Ref', 'Alt', \"control_snp\"]\n",
    "control_merged.columns = _\n",
    "control_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load Case data\n",
    "case_merged = pd.read_csv(\"30_case_mearged.csv\", dtype= \"object\", low_memory= False)\n",
    "_ = ['Chr', 'Start', 'Ref', 'Alt', \"Case_snp\"]\n",
    "case_merged.columns = _\n",
    "case_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge Case and Control\n",
    "#pd.merge(case_merged,control_merged , on = open_cols, how = \"right\").to_csv(\"Case_Control_60.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = pd.merge(case_merged,control_merged , on = open_cols, how = \"right\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#merged[merged.Case_snp >= 25]\n",
    "merged.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.Case_snp = merged.Case_snp.astype(\"float64\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged[(merged.Case_snp >= 26) & (merged.control_snp <14)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = \"1119\"\n",
    "df = pd.read_csv(_+\"/\"+_+\"_snv.csv\", nrows = 50, usecols = open_cols )\n",
    "df[\"SNP\"] = 1\n",
    "case_merged = df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "case_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = \"1093\"\n",
    "df_ = pd.read_csv(_+\"/\"+_+\"_snv.csv\", nrows = 10, usecols = open_cols )\n",
    "df_[\"SNP_\"] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#case_merged = pd.\n",
    "case_merged = pd.merge(df, df_, on=open_cols, how=\"outer\")\n",
    "case_merged.fillna(0, inplace = True)\n",
    "case_merged[\"SNP\"] =  case_merged.SNP + case_merged.SNP_\n",
    "case_merged.drop([\"SNP_\"], axis = 1, inplace = True)\n",
    "case_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "case_merged = pd.merge(df, df_, on=open_cols, how=\"outer\")\n",
    "case_merged[\"SNP\"] = case_merged[\"SNP_x\"]+ case_merged[\"SNP_y\"]\n",
    "case_merged.drop(['SNP_x',\"SNP_y\"], axis = 1, inplace = True)\n",
    "\n",
    "case_merged = pd.merge(case_merged, df_, on=merged_cols, how=\"outer\")\n",
    "#case_merged[\"SNP\"] = case_merged[\"SNP\"]+ case_merged[\"SNP_x\"]\n",
    "#case_merged.drop(['SNP_x',\"SNP_y\"], axis = 1, inplace = True)\n",
    "case_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lets open Indel Files from Intervar and merge as a single Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"1093/1093_snv.csv\", low_memory= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df = df[df.Func_refGene.isin([\"exonic\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Ref_Gene</th>\n",
       "      <th>Func_refGene</th>\n",
       "      <th>avsnp147</th>\n",
       "      <th>Clinvar</th>\n",
       "      <th>InterVar</th>\n",
       "      <th>OMIM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>191249</th>\n",
       "      <td>1</td>\n",
       "      <td>110149039</td>\n",
       "      <td>110149039</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GNAT2</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs745308973</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>139340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3561175</th>\n",
       "      <td>11</td>\n",
       "      <td>64361219</td>\n",
       "      <td>64361219</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SLC22A12</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs121907892</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>607096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3594510</th>\n",
       "      <td>11</td>\n",
       "      <td>85365183</td>\n",
       "      <td>85365183</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TMEM126A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs121434508</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>612988</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr      Start        End Ref Alt  Ref_Gene Func_refGene     avsnp147             Clinvar    InterVar    OMIM\n",
       "191249    1  110149039  110149039   G   A     GNAT2       exonic  rs745308973  Likely_pathogenic   Pathogenic  139340\n",
       "3561175  11   64361219   64361219   G   A  SLC22A12       exonic  rs121907892         Pathogenic   Pathogenic  607096\n",
       "3594510  11   85365183   85365183   C   T  TMEM126A       exonic  rs121434508         Pathogenic   Pathogenic  612988"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.InterVar.isin([\"Pathogenic\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ncRNA_exonic', 'ncRNA_intronic', 'intergenic', 'exonic',\n",
       "       'downstream', 'upstream', 'intronic', 'UTR3', 'UTR5',\n",
       "       'upstream;downstream', 'ncRNA_splicing', 'splicing',\n",
       "       'exonic;splicing', 'ncRNA_exonic;splicing', 'UTR5;UTR3'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.Func_refGene.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Ref_Gene</th>\n",
       "      <th>Func_refGene</th>\n",
       "      <th>avsnp147</th>\n",
       "      <th>Clinvar</th>\n",
       "      <th>InterVar</th>\n",
       "      <th>OMIM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>320757</th>\n",
       "      <td>1</td>\n",
       "      <td>198867678</td>\n",
       "      <td>198867678</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>MIR181A1HG</td>\n",
       "      <td>ncRNA_intronic</td>\n",
       "      <td>rs60639710</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Benign</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320758</th>\n",
       "      <td>1</td>\n",
       "      <td>198868084</td>\n",
       "      <td>198868084</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>MIR181A1HG</td>\n",
       "      <td>ncRNA_intronic</td>\n",
       "      <td>rs10800597</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Benign</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320759</th>\n",
       "      <td>1</td>\n",
       "      <td>198869514</td>\n",
       "      <td>198869514</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>MIR181A1HG</td>\n",
       "      <td>ncRNA_intronic</td>\n",
       "      <td>rs10800598</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Benign</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1242327</th>\n",
       "      <td>4</td>\n",
       "      <td>6295693</td>\n",
       "      <td>6295693</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>WFS1</td>\n",
       "      <td>intronic</td>\n",
       "      <td>rs6446482</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Benign</td>\n",
       "      <td>606201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1630820</th>\n",
       "      <td>4</td>\n",
       "      <td>187158034</td>\n",
       "      <td>187158034</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>KLKB1</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs3733402</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Benign</td>\n",
       "      <td>229000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1921435</th>\n",
       "      <td>5</td>\n",
       "      <td>150227998</td>\n",
       "      <td>150227998</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>IRGM</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs10065172</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Benign</td>\n",
       "      <td>608212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2683581</th>\n",
       "      <td>8</td>\n",
       "      <td>11606312</td>\n",
       "      <td>11606312</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>GATA4</td>\n",
       "      <td>intronic</td>\n",
       "      <td>rs3735819</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Benign</td>\n",
       "      <td>600576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3271703</th>\n",
       "      <td>10</td>\n",
       "      <td>54531235</td>\n",
       "      <td>54531235</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>MBL2</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs1800450</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Benign</td>\n",
       "      <td>154545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3530258</th>\n",
       "      <td>11</td>\n",
       "      <td>48145375</td>\n",
       "      <td>48145375</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>PTPRJ</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs1566734</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Benign</td>\n",
       "      <td>600925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3561175</th>\n",
       "      <td>11</td>\n",
       "      <td>64361219</td>\n",
       "      <td>64361219</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SLC22A12</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs121907892</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>607096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3594510</th>\n",
       "      <td>11</td>\n",
       "      <td>85365183</td>\n",
       "      <td>85365183</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TMEM126A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs121434508</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>612988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5434738</th>\n",
       "      <td>X</td>\n",
       "      <td>105278361</td>\n",
       "      <td>105278361</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SERPINA7</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs1804495</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Benign</td>\n",
       "      <td>314200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5448779</th>\n",
       "      <td>X</td>\n",
       "      <td>119760629</td>\n",
       "      <td>119760629</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C1GALT1C1</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs17261572</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Benign</td>\n",
       "      <td>300611</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr      Start        End Ref Alt    Ref_Gene    Func_refGene     avsnp147      Clinvar    InterVar    OMIM\n",
       "320757    1  198867678  198867678   G   T  MIR181A1HG  ncRNA_intronic   rs60639710  Pathogenic       Benign       .\n",
       "320758    1  198868084  198868084   G   A  MIR181A1HG  ncRNA_intronic   rs10800597  Pathogenic       Benign       .\n",
       "320759    1  198869514  198869514   T   C  MIR181A1HG  ncRNA_intronic   rs10800598  Pathogenic       Benign       .\n",
       "1242327   4    6295693    6295693   C   G        WFS1        intronic    rs6446482  Pathogenic       Benign  606201\n",
       "1630820   4  187158034  187158034   G   A       KLKB1          exonic    rs3733402  Pathogenic       Benign  229000\n",
       "1921435   5  150227998  150227998   C   T        IRGM          exonic   rs10065172  Pathogenic       Benign  608212\n",
       "2683581   8   11606312   11606312   T   C       GATA4        intronic    rs3735819  Pathogenic       Benign  600576\n",
       "3271703  10   54531235   54531235   C   T        MBL2          exonic    rs1800450  Pathogenic       Benign  154545\n",
       "3530258  11   48145375   48145375   A   C       PTPRJ          exonic    rs1566734  Pathogenic       Benign  600925\n",
       "3561175  11   64361219   64361219   G   A    SLC22A12          exonic  rs121907892  Pathogenic   Pathogenic  607096\n",
       "3594510  11   85365183   85365183   C   T    TMEM126A          exonic  rs121434508  Pathogenic   Pathogenic  612988\n",
       "5434738   X  105278361  105278361   C   A    SERPINA7          exonic    rs1804495  Pathogenic       Benign  314200\n",
       "5448779   X  119760629  119760629   A   T   C1GALT1C1          exonic   rs17261572  Pathogenic       Benign  300611"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.Clinvar.isin([\"Pathogenic \"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Ref_Gene</th>\n",
       "      <th>Func_refGene</th>\n",
       "      <th>avsnp147</th>\n",
       "      <th>Clinvar</th>\n",
       "      <th>InterVar</th>\n",
       "      <th>OMIM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>13868</td>\n",
       "      <td>13868</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>rs796086906</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>13868</td>\n",
       "      <td>13868</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>LOC102725121</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>rs796086906</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>15820</td>\n",
       "      <td>15820</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>rs2691315</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>1</td>\n",
       "      <td>135465</td>\n",
       "      <td>135465</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>LOC729737</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>rs375289924</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>1</td>\n",
       "      <td>135982</td>\n",
       "      <td>135982</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>LOC729737</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>rs62642131</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5491749</th>\n",
       "      <td>X</td>\n",
       "      <td>152864826</td>\n",
       "      <td>152864826</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>LOC105373383</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>rs35908555</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5491750</th>\n",
       "      <td>X</td>\n",
       "      <td>152865326</td>\n",
       "      <td>152865326</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>LOC105373383</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>rs11156603</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5492077</th>\n",
       "      <td>X</td>\n",
       "      <td>153149715</td>\n",
       "      <td>153149715</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>LCA10</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>rs6643650</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5493207</th>\n",
       "      <td>M</td>\n",
       "      <td>2354</td>\n",
       "      <td>2354</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>RNR2</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>.</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>180451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5493208</th>\n",
       "      <td>M</td>\n",
       "      <td>2485</td>\n",
       "      <td>2485</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>RNR2</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>.</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>180451</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14349 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr      Start        End Ref Alt      Ref_Gene  Func_refGene     avsnp147 Clinvar                InterVar    OMIM\n",
       "0         1      13868      13868   A   G       DDX11L1  ncRNA_exonic  rs796086906    UNK                   Benign       .\n",
       "1         1      13868      13868   A   G  LOC102725121  ncRNA_exonic  rs796086906    UNK                   Benign       .\n",
       "7         1      15820      15820   G   T        WASH7P  ncRNA_exonic    rs2691315    UNK                   Benign       .\n",
       "150       1     135465     135465   C   T     LOC729737  ncRNA_exonic  rs375289924    UNK                   Benign       .\n",
       "151       1     135982     135982   A   G     LOC729737  ncRNA_exonic   rs62642131    UNK                   Benign       .\n",
       "...      ..        ...        ...  ..  ..           ...           ...          ...     ...                     ...     ...\n",
       "5491749   X  152864826  152864826   G   C  LOC105373383  ncRNA_exonic   rs35908555    UNK                   Benign       .\n",
       "5491750   X  152865326  152865326   C   T  LOC105373383  ncRNA_exonic   rs11156603    UNK                   Benign       .\n",
       "5492077   X  153149715  153149715   C   G         LCA10  ncRNA_exonic    rs6643650    UNK                   Benign       .\n",
       "5493207   M       2354       2354   C   T          RNR2  ncRNA_exonic            .    UNK   Uncertain_significance  180451\n",
       "5493208   M       2485       2485   C   T          RNR2  ncRNA_exonic            .    UNK   Uncertain_significance  180451\n",
       "\n",
       "[14349 rows x 11 columns]"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"Func_refGene\"] == \"ncRNA_exonic\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Ref_Gene</th>\n",
       "      <th>Func_refGene</th>\n",
       "      <th>avsnp147</th>\n",
       "      <th>Clinvar</th>\n",
       "      <th>InterVar</th>\n",
       "      <th>OMIM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>1</td>\n",
       "      <td>69511</td>\n",
       "      <td>69511</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs2691305</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>1</td>\n",
       "      <td>69897</td>\n",
       "      <td>69897</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs200676709</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>1</td>\n",
       "      <td>877831</td>\n",
       "      <td>877831</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs6672356</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>616765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>1</td>\n",
       "      <td>879317</td>\n",
       "      <td>879317</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs7523549</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>616765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>635</th>\n",
       "      <td>1</td>\n",
       "      <td>881627</td>\n",
       "      <td>881627</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs2272757</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>610770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5492384</th>\n",
       "      <td>X</td>\n",
       "      <td>153695908</td>\n",
       "      <td>153695908</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>PLXNA3</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs60361517</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>300022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5492385</th>\n",
       "      <td>X</td>\n",
       "      <td>153696247</td>\n",
       "      <td>153696247</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>PLXNA3</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs5987266</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>300022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5492508</th>\n",
       "      <td>X</td>\n",
       "      <td>153994596</td>\n",
       "      <td>153994596</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>DKC1</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs2728532</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Benign</td>\n",
       "      <td>300126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5492516</th>\n",
       "      <td>X</td>\n",
       "      <td>154020114</td>\n",
       "      <td>154020114</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>MPP1</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs1126762</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>305360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5492646</th>\n",
       "      <td>X</td>\n",
       "      <td>154456747</td>\n",
       "      <td>154456747</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>VBP1</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs572013</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Benign</td>\n",
       "      <td>300133</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22693 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr      Start        End Ref Alt Ref_Gene Func_refGene     avsnp147  Clinvar InterVar    OMIM\n",
       "64        1      69511      69511   A   G    OR4F5       exonic    rs2691305     UNK    Benign       .\n",
       "65        1      69897      69897   T   C    OR4F5       exonic  rs200676709     UNK    Benign       .\n",
       "627       1     877831     877831   T   C   SAMD11       exonic    rs6672356     UNK    Benign  616765\n",
       "628       1     879317     879317   C   T   SAMD11       exonic    rs7523549     UNK    Benign  616765\n",
       "635       1     881627     881627   G   A    NOC2L       exonic    rs2272757     UNK    Benign  610770\n",
       "...      ..        ...        ...  ..  ..      ...          ...          ...      ...      ...     ...\n",
       "5492384   X  153695908  153695908   G   T   PLXNA3       exonic   rs60361517     UNK    Benign  300022\n",
       "5492385   X  153696247  153696247   G   C   PLXNA3       exonic    rs5987266     UNK    Benign  300022\n",
       "5492508   X  153994596  153994596   G   T     DKC1       exonic    rs2728532  Benign    Benign  300126\n",
       "5492516   X  154020114  154020114   C   A     MPP1       exonic    rs1126762     UNK    Benign  305360\n",
       "5492646   X  154456747  154456747   A   G     VBP1       exonic     rs572013     UNK    Benign  300133\n",
       "\n",
       "[22693 rows x 11 columns]"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df = df.loc[:,['Chr', 'Start',  'Ref', 'Alt', 'Ref_Gene', 'Func_refGene', 'avsnp147', 'Clinvar', 'InterVar', 'OMIM']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = df.head(100).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Ref_Gene</th>\n",
       "      <th>Func_refGene</th>\n",
       "      <th>avsnp147</th>\n",
       "      <th>Clinvar</th>\n",
       "      <th>InterVar</th>\n",
       "      <th>OMIM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>191249</th>\n",
       "      <td>1</td>\n",
       "      <td>110149039</td>\n",
       "      <td>110149039</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GNAT2</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs745308973</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>139340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3561175</th>\n",
       "      <td>11</td>\n",
       "      <td>64361219</td>\n",
       "      <td>64361219</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SLC22A12</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs121907892</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>607096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3594510</th>\n",
       "      <td>11</td>\n",
       "      <td>85365183</td>\n",
       "      <td>85365183</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TMEM126A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs121434508</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>612988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3731267</th>\n",
       "      <td>12</td>\n",
       "      <td>18865906</td>\n",
       "      <td>18865906</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>PLCZ1</td>\n",
       "      <td>exonic</td>\n",
       "      <td>.</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>608075</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr      Start        End Ref Alt  Ref_Gene Func_refGene     avsnp147             Clinvar           InterVar    OMIM\n",
       "191249    1  110149039  110149039   G   A     GNAT2       exonic  rs745308973  Likely_pathogenic          Pathogenic  139340\n",
       "3561175  11   64361219   64361219   G   A  SLC22A12       exonic  rs121907892         Pathogenic          Pathogenic  607096\n",
       "3594510  11   85365183   85365183   C   T  TMEM126A       exonic  rs121434508         Pathogenic          Pathogenic  612988\n",
       "3731267  12   18865906   18865906   C   A     PLCZ1       exonic            .                UNK   Likely_pathogenic  608075"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.InterVar.isin(b)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "intergenic               3894404\n",
       "intronic                 1265068\n",
       "ncRNA_intronic            239854\n",
       "UTR3                       32627\n",
       "downstream                 24154\n",
       "exonic                     22875\n",
       "upstream                   22728\n",
       "ncRNA_exonic               14590\n",
       "UTR5                        5869\n",
       "upstream;downstream         1703\n",
       "splicing                      97\n",
       "ncRNA_splicing                84\n",
       "UTR5;UTR3                     32\n",
       "exonic;splicing               18\n",
       "ncRNA_exonic;splicing         12\n",
       "Name: Func_refGene, dtype: int64"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df.Func_refGene.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### Intergenic Genes --> are junk dna and non coding regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "preterm_genes = [\"WNT4\", \"EFB1\", \"TEKT3\", \"TGFB1\", \"AGTR2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df[df.Ref_Gene.isin(preterm_genes)]\n",
    "a = [\"UNK \", \"Benign\",\"Uncertain_significance\"]\n",
    "b= [\"Likely_pathogenic\", \"Pathogenic\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Ref_Gene</th>\n",
       "      <th>Func_refGene</th>\n",
       "      <th>avsnp147</th>\n",
       "      <th>Clinvar</th>\n",
       "      <th>InterVar</th>\n",
       "      <th>OMIM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>191249</th>\n",
       "      <td>1</td>\n",
       "      <td>110149039</td>\n",
       "      <td>110149039</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GNAT2</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs745308973</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>139340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3561175</th>\n",
       "      <td>11</td>\n",
       "      <td>64361219</td>\n",
       "      <td>64361219</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SLC22A12</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs121907892</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>607096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3594510</th>\n",
       "      <td>11</td>\n",
       "      <td>85365183</td>\n",
       "      <td>85365183</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TMEM126A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>rs121434508</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>612988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3731267</th>\n",
       "      <td>12</td>\n",
       "      <td>18865906</td>\n",
       "      <td>18865906</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>PLCZ1</td>\n",
       "      <td>exonic</td>\n",
       "      <td>.</td>\n",
       "      <td>UNK</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>608075</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr      Start        End Ref Alt  Ref_Gene Func_refGene     avsnp147             Clinvar           InterVar    OMIM\n",
       "191249    1  110149039  110149039   G   A     GNAT2       exonic  rs745308973  Likely_pathogenic          Pathogenic  139340\n",
       "3561175  11   64361219   64361219   G   A  SLC22A12       exonic  rs121907892         Pathogenic          Pathogenic  607096\n",
       "3594510  11   85365183   85365183   C   T  TMEM126A       exonic  rs121434508         Pathogenic          Pathogenic  612988\n",
       "3731267  12   18865906   18865906   C   A     PLCZ1       exonic            .                UNK   Likely_pathogenic  608075"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df.InterVar.isin(b))]# & (df.InterVar.isin(preterm_genes))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[df.Func_refGene == \"exonic\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Ref_Gene</th>\n",
       "      <th>Func_refGene</th>\n",
       "      <th>avsnp147</th>\n",
       "      <th>Clinvar</th>\n",
       "      <th>InterVar</th>\n",
       "      <th>OMIM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chr, Start, End, Ref, Alt, Ref_Gene, Func_refGene, avsnp147, Clinvar, InterVar, OMIM]\n",
       "Index: []"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df.Ref_Gene.value_counts()\n",
    "import seaborn as sns\n",
    "#sns.distplot(df.Ref_Gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "## For distribution Plot We lets import original vcf with Frequency\n",
    "#df = pd.read_csv(\"../old/910.PASS.vcf\", skiprows=81, nrows=100, sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df['default'] = df.default.apply(lambda x : x.split(\",\")[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Distribution Plot\n",
    "#sns.distplot(df.default)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### From Intervar we want Only before  PVS1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\krish\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3063: DtypeWarning: Columns (4) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "df_ = pd.read_csv(\"Merged_vcf/Final_mergedf_case_con_vcf_og_pass.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Repeated_Case</th>\n",
       "      <th>MAX_Case</th>\n",
       "      <th>Repeated_Control</th>\n",
       "      <th>MAX_Control</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10146</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>Counter({'0/1': 1})</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10427</td>\n",
       "      <td>A</td>\n",
       "      <td>AC</td>\n",
       "      <td>Counter({'0/1': 1})</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Counter({'0/1': 1})</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10439</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>Counter({'0/1': 1})</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Counter({'0/1': 1})</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10492</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Counter({'0/1': 2})</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Counter({'0/1': 2})</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10583</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Counter({'0/1': 1})</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Counter({'0/1': 1})</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM    POS REF ALT        Repeated_Case  MAX_Case     Repeated_Control  MAX_Control\n",
       "0  chr1  10146  AC   A  Counter({'0/1': 1})       1.0                  NaN          NaN\n",
       "1  chr1  10427   A  AC  Counter({'0/1': 1})       1.0  Counter({'0/1': 1})          1.0\n",
       "2  chr1  10439  AC   A  Counter({'0/1': 1})       1.0  Counter({'0/1': 1})          1.0\n",
       "3  chr1  10492   C   T  Counter({'0/1': 2})       2.0  Counter({'0/1': 2})          2.0\n",
       "4  chr1  10583   G   A  Counter({'0/1': 1})       1.0  Counter({'0/1': 1})          1.0"
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(490, 8)"
      ]
     },
     "execution_count": 259,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_[(df_[\"MAX_Case\"]>20) & (df_[\"MAX_Control\"]<10)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(487, 8)"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_[(df_[\"MAX_Case\"]<10) & (df_[\"MAX_Control\"]>20)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "nbTranslate": {
   "displayLangs": [
    "*"
   ],
   "hotkey": "alt-t",
   "langInMainMenu": true,
   "sourceLang": "en",
   "targetLang": "fr",
   "useGoogleTranslate": true
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
