{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running doctests...\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Not within Biopython source tree: 'C:\\\\Users\\\\user\\\\Desktop\\\\Git_Hub'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-2-9d591c437eee>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m   1530\u001b[0m \u001b[1;32mif\u001b[0m \u001b[0m__name__\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m\"__main__\"\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1531\u001b[0m     \u001b[1;32mfrom\u001b[0m \u001b[0mBio\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_utils\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mrun_doctest\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1532\u001b[1;33m     \u001b[0mrun_doctest\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\_utils.py\u001b[0m in \u001b[0;36mrun_doctest\u001b[1;34m(target_dir, *args, **kwargs)\u001b[0m\n\u001b[0;32m    129\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Running doctests...\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    130\u001b[0m     \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 131\u001b[1;33m         \u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchdir\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfind_test_dir\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtarget_dir\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    132\u001b[0m         \u001b[0mdoctest\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtestmod\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    133\u001b[0m     \u001b[1;32mfinally\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\_utils.py\u001b[0m in \u001b[0;36mfind_test_dir\u001b[1;34m(start_dir)\u001b[0m\n\u001b[0;32m    112\u001b[0m         \u001b[0mtarget\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnew\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    113\u001b[0m     raise ValueError(\"Not within Biopython source tree: %r\" %\n\u001b[1;32m--> 114\u001b[1;33m                      os.path.abspath(start_dir))\n\u001b[0m\u001b[0;32m    115\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    116\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: Not within Biopython source tree: 'C:\\\\Users\\\\user\\\\Desktop\\\\Git_Hub'"
     ]
    }
   ],
   "source": [
    "# Copyright 2000, 2004 by Brad Chapman.\n",
    "# Revisions copyright 2010-2013, 2015-2018 by Peter Cock.\n",
    "# All rights reserved.\n",
    "#\n",
    "# This file is part of the Biopython distribution and governed by your\n",
    "# choice of the \"Biopython License Agreement\" or the \"BSD 3-Clause License\".\n",
    "# Please see the LICENSE file that should have been included as part of this\n",
    "# package.\n",
    "\"\"\"Code for dealing with sequence alignments.\n",
    "\n",
    "One of the most important things in this module is the MultipleSeqAlignment\n",
    "class, used in the Bio.AlignIO module.\n",
    "\n",
    "\"\"\"\n",
    "from __future__ import print_function\n",
    "\n",
    "import sys  # Only needed to check if we are using Python 2 or 3\n",
    "\n",
    "from Bio._py3k import raise_from\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord, _RestrictedDict\n",
    "from Bio import Alphabet\n",
    "\n",
    "from Bio.Align import _aligners\n",
    "# Import errors may occur here if a compiled aligners.c file\n",
    "# (_aligners.pyd or _aligners.so) is missing or if the user is\n",
    "# importing from within the Biopython source tree, see PR #2007:\n",
    "# https://github.com/biopython/biopython/pull/2007\n",
    "\n",
    "\n",
    "class MultipleSeqAlignment(object):\n",
    "    \"\"\"Represents a classical multiple sequence alignment (MSA).\n",
    "\n",
    "    By this we mean a collection of sequences (usually shown as rows) which\n",
    "    are all the same length (usually with gap characters for insertions or\n",
    "    padding). The data can then be regarded as a matrix of letters, with well\n",
    "    defined columns.\n",
    "\n",
    "    You would typically create an MSA by loading an alignment file with the\n",
    "    AlignIO module:\n",
    "\n",
    "    >>> from Bio import AlignIO\n",
    "    >>> align = AlignIO.read(\"Clustalw/opuntia.aln\", \"clustal\")\n",
    "    >>> print(align)\n",
    "    SingleLetterAlphabet() alignment with 7 rows and 156 columns\n",
    "    TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191\n",
    "    TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191\n",
    "    TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191\n",
    "    TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191\n",
    "    TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191\n",
    "    TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191\n",
    "    TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191\n",
    "\n",
    "    In some respects you can treat these objects as lists of SeqRecord objects,\n",
    "    each representing a row of the alignment. Iterating over an alignment gives\n",
    "    the SeqRecord object for each row:\n",
    "\n",
    "    >>> len(align)\n",
    "    7\n",
    "    >>> for record in align:\n",
    "    ...     print(\"%s %i\" % (record.id, len(record)))\n",
    "    ...\n",
    "    gi|6273285|gb|AF191659.1|AF191 156\n",
    "    gi|6273284|gb|AF191658.1|AF191 156\n",
    "    gi|6273287|gb|AF191661.1|AF191 156\n",
    "    gi|6273286|gb|AF191660.1|AF191 156\n",
    "    gi|6273290|gb|AF191664.1|AF191 156\n",
    "    gi|6273289|gb|AF191663.1|AF191 156\n",
    "    gi|6273291|gb|AF191665.1|AF191 156\n",
    "\n",
    "    You can also access individual rows as SeqRecord objects via their index:\n",
    "\n",
    "    >>> print(align[0].id)\n",
    "    gi|6273285|gb|AF191659.1|AF191\n",
    "    >>> print(align[-1].id)\n",
    "    gi|6273291|gb|AF191665.1|AF191\n",
    "\n",
    "    And extract columns as strings:\n",
    "\n",
    "    >>> print(align[:, 1])\n",
    "    AAAAAAA\n",
    "\n",
    "    Or, take just the first ten columns as a sub-alignment:\n",
    "\n",
    "    >>> print(align[:, :10])\n",
    "    SingleLetterAlphabet() alignment with 7 rows and 10 columns\n",
    "    TATACATTAA gi|6273285|gb|AF191659.1|AF191\n",
    "    TATACATTAA gi|6273284|gb|AF191658.1|AF191\n",
    "    TATACATTAA gi|6273287|gb|AF191661.1|AF191\n",
    "    TATACATAAA gi|6273286|gb|AF191660.1|AF191\n",
    "    TATACATTAA gi|6273290|gb|AF191664.1|AF191\n",
    "    TATACATTAA gi|6273289|gb|AF191663.1|AF191\n",
    "    TATACATTAA gi|6273291|gb|AF191665.1|AF191\n",
    "\n",
    "    Combining this alignment slicing with alignment addition allows you to\n",
    "    remove a section of the alignment. For example, taking just the first\n",
    "    and last ten columns:\n",
    "\n",
    "    >>> print(align[:, :10] + align[:, -10:])\n",
    "    SingleLetterAlphabet() alignment with 7 rows and 20 columns\n",
    "    TATACATTAAGTGTACCAGA gi|6273285|gb|AF191659.1|AF191\n",
    "    TATACATTAAGTGTACCAGA gi|6273284|gb|AF191658.1|AF191\n",
    "    TATACATTAAGTGTACCAGA gi|6273287|gb|AF191661.1|AF191\n",
    "    TATACATAAAGTGTACCAGA gi|6273286|gb|AF191660.1|AF191\n",
    "    TATACATTAAGTGTACCAGA gi|6273290|gb|AF191664.1|AF191\n",
    "    TATACATTAAGTATACCAGA gi|6273289|gb|AF191663.1|AF191\n",
    "    TATACATTAAGTGTACCAGA gi|6273291|gb|AF191665.1|AF191\n",
    "\n",
    "    Note - This object replaced the older Alignment object defined in module\n",
    "    Bio.Align.Generic but is not fully backwards compatible with it.\n",
    "\n",
    "    Note - This object does NOT attempt to model the kind of alignments used\n",
    "    in next generation sequencing with multiple sequencing reads which are\n",
    "    much shorter than the alignment, and where there is usually a consensus or\n",
    "    reference sequence with special status.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, records, alphabet=None,\n",
    "                 annotations=None, column_annotations=None):\n",
    "        \"\"\"Initialize a new MultipleSeqAlignment object.\n",
    "\n",
    "        Arguments:\n",
    "         - records - A list (or iterator) of SeqRecord objects, whose\n",
    "                     sequences are all the same length.  This may be an be an\n",
    "                     empty list.\n",
    "         - alphabet - The alphabet for the whole alignment, typically a gapped\n",
    "                      alphabet, which should be a super-set of the individual\n",
    "                      record alphabets.  If omitted, a consensus alphabet is\n",
    "                      used.\n",
    "         - annotations - Information about the whole alignment (dictionary).\n",
    "         - column_annotations - Per column annotation (restricted dictionary).\n",
    "                      This holds Python sequences (lists, strings, tuples)\n",
    "                      whose length matches the number of columns. A typical\n",
    "                      use would be a secondary structure consensus string.\n",
    "\n",
    "        You would normally load a MSA from a file using Bio.AlignIO, but you\n",
    "        can do this from a list of SeqRecord objects too:\n",
    "\n",
    "        >>> from Bio.Alphabet import generic_dna\n",
    "        >>> from Bio.Seq import Seq\n",
    "        >>> from Bio.SeqRecord import SeqRecord\n",
    "        >>> from Bio.Align import MultipleSeqAlignment\n",
    "        >>> a = SeqRecord(Seq(\"AAAACGT\", generic_dna), id=\"Alpha\")\n",
    "        >>> b = SeqRecord(Seq(\"AAA-CGT\", generic_dna), id=\"Beta\")\n",
    "        >>> c = SeqRecord(Seq(\"AAAAGGT\", generic_dna), id=\"Gamma\")\n",
    "        >>> align = MultipleSeqAlignment([a, b, c],\n",
    "        ...                              annotations={\"tool\": \"demo\"},\n",
    "        ...                              column_annotations={\"stats\": \"CCCXCCC\"})\n",
    "        >>> print(align)\n",
    "        DNAAlphabet() alignment with 3 rows and 7 columns\n",
    "        AAAACGT Alpha\n",
    "        AAA-CGT Beta\n",
    "        AAAAGGT Gamma\n",
    "        >>> align.annotations\n",
    "        {'tool': 'demo'}\n",
    "        >>> align.column_annotations\n",
    "        {'stats': 'CCCXCCC'}\n",
    "        \"\"\"\n",
    "        if alphabet is not None:\n",
    "            if not isinstance(alphabet, (Alphabet.Alphabet, Alphabet.AlphabetEncoder)):\n",
    "                raise ValueError(\"Invalid alphabet argument\")\n",
    "            self._alphabet = alphabet\n",
    "        else:\n",
    "            # Default while we add sequences, will take a consensus later\n",
    "            self._alphabet = Alphabet.single_letter_alphabet\n",
    "\n",
    "        self._records = []\n",
    "        if records:\n",
    "            self.extend(records)\n",
    "            if alphabet is None:\n",
    "                # No alphabet was given, take a consensus alphabet\n",
    "                self._alphabet = Alphabet._consensus_alphabet(rec.seq.alphabet for\n",
    "                                                              rec in self._records\n",
    "                                                              if rec.seq is not None)\n",
    "\n",
    "        # Annotations about the whole alignment\n",
    "        if annotations is None:\n",
    "            annotations = {}\n",
    "        elif not isinstance(annotations, dict):\n",
    "            raise TypeError(\"annotations argument should be a dict\")\n",
    "        self.annotations = annotations\n",
    "\n",
    "        # Annotations about each colum of the alignment\n",
    "        if column_annotations is None:\n",
    "            column_annotations = {}\n",
    "        # Handle this via the property set function which will validate it\n",
    "        self.column_annotations = column_annotations\n",
    "\n",
    "    def _set_per_column_annotations(self, value):\n",
    "        if not isinstance(value, dict):\n",
    "            raise TypeError(\"The per-column-annotations should be a \"\n",
    "                            \"(restricted) dictionary.\")\n",
    "        # Turn this into a restricted-dictionary (and check the entries)\n",
    "        if len(self):\n",
    "            # Use the standard method to get the length\n",
    "            expected_length = self.get_alignment_length()\n",
    "            self._per_col_annotations = _RestrictedDict(length=expected_length)\n",
    "            self._per_col_annotations.update(value)\n",
    "        else:\n",
    "            # Bit of a problem case... number of columns is undefined\n",
    "            self._per_col_annotations = None\n",
    "            if value:\n",
    "                raise ValueError(\"Can't set per-column-annotations without an alignment\")\n",
    "\n",
    "    def _get_per_column_annotations(self):\n",
    "        if self._per_col_annotations is None:\n",
    "            # This happens if empty at initialisation\n",
    "            if len(self):\n",
    "                # Use the standard method to get the length\n",
    "                expected_length = self.get_alignment_length()\n",
    "            else:\n",
    "                # Should this raise an exception? Compare SeqRecord behaviour...\n",
    "                expected_length = 0\n",
    "            self._per_col_annotations = _RestrictedDict(length=expected_length)\n",
    "        return self._per_col_annotations\n",
    "\n",
    "    column_annotations = property(\n",
    "        fget=_get_per_column_annotations,\n",
    "        fset=_set_per_column_annotations,\n",
    "        doc=\"\"\"Dictionary of per-letter-annotation for the sequence.\"\"\")\n",
    "\n",
    "    def _str_line(self, record, length=50):\n",
    "        \"\"\"Return a truncated string representation of a SeqRecord (PRIVATE).\n",
    "\n",
    "        This is a PRIVATE function used by the __str__ method.\n",
    "        \"\"\"\n",
    "        if record.seq.__class__.__name__ == \"CodonSeq\":\n",
    "            if len(record.seq) <= length:\n",
    "                return \"%s %s\" % (record.seq, record.id)\n",
    "            else:\n",
    "                return \"%s...%s %s\" \\\n",
    "                    % (record.seq[:length - 3], record.seq[-3:], record.id)\n",
    "        else:\n",
    "            if len(record.seq) <= length:\n",
    "                return \"%s %s\" % (record.seq, record.id)\n",
    "            else:\n",
    "                return \"%s...%s %s\" \\\n",
    "                    % (record.seq[:length - 6], record.seq[-3:], record.id)\n",
    "\n",
    "    def __str__(self):\n",
    "        \"\"\"Return a multi-line string summary of the alignment.\n",
    "\n",
    "        This output is intended to be readable, but large alignments are\n",
    "        shown truncated.  A maximum of 20 rows (sequences) and 50 columns\n",
    "        are shown, with the record identifiers.  This should fit nicely on a\n",
    "        single screen. e.g.\n",
    "\n",
    "        >>> from Bio.Alphabet import IUPAC, Gapped\n",
    "        >>> from Bio.Align import MultipleSeqAlignment\n",
    "        >>> align = MultipleSeqAlignment([], Gapped(IUPAC.unambiguous_dna, \"-\"))\n",
    "        >>> align.add_sequence(\"Alpha\", \"ACTGCTAGCTAG\")\n",
    "        >>> align.add_sequence(\"Beta\",  \"ACT-CTAGCTAG\")\n",
    "        >>> align.add_sequence(\"Gamma\", \"ACTGCTAGATAG\")\n",
    "        >>> print(align)\n",
    "        Gapped(IUPACUnambiguousDNA(), '-') alignment with 3 rows and 12 columns\n",
    "        ACTGCTAGCTAG Alpha\n",
    "        ACT-CTAGCTAG Beta\n",
    "        ACTGCTAGATAG Gamma\n",
    "\n",
    "        See also the alignment's format method.\n",
    "        \"\"\"\n",
    "        rows = len(self._records)\n",
    "        lines = [\"%s alignment with %i rows and %i columns\"\n",
    "                 % (str(self._alphabet), rows, self.get_alignment_length())]\n",
    "        if rows <= 20:\n",
    "            lines.extend(self._str_line(rec) for rec in self._records)\n",
    "        else:\n",
    "            lines.extend(self._str_line(rec) for rec in self._records[:18])\n",
    "            lines.append(\"...\")\n",
    "            lines.append(self._str_line(self._records[-1]))\n",
    "        return \"\\n\".join(lines)\n",
    "\n",
    "    def __repr__(self):\n",
    "        \"\"\"Return a representation of the object for debugging.\n",
    "\n",
    "        The representation cannot be used with eval() to recreate the object,\n",
    "        which is usually possible with simple python ojects.  For example:\n",
    "\n",
    "        <Bio.Align.MultipleSeqAlignment instance (2 records of length 14,\n",
    "        SingleLetterAlphabet()) at a3c184c>\n",
    "\n",
    "        The hex string is the memory address of the object, see help(id).\n",
    "        This provides a simple way to visually distinguish alignments of\n",
    "        the same size.\n",
    "        \"\"\"\n",
    "        # A doctest for __repr__ would be nice, but __class__ comes out differently\n",
    "        # if run via the __main__ trick.\n",
    "        return \"<%s instance (%i records of length %i, %s) at %x>\" % \\\n",
    "            (self.__class__, len(self._records),\n",
    "             self.get_alignment_length(), repr(self._alphabet), id(self))\n",
    "        # This version is useful for doing eval(repr(alignment)),\n",
    "        # but it can be VERY long:\n",
    "        # return \"%s(%s, %s)\" \\\n",
    "        #       % (self.__class__, repr(self._records), repr(self._alphabet))\n",
    "\n",
    "    def format(self, format):\n",
    "        \"\"\"Return the alignment as a string in the specified file format.\n",
    "\n",
    "        The format should be a lower case string supported as an output\n",
    "        format by Bio.AlignIO (such as \"fasta\", \"clustal\", \"phylip\",\n",
    "        \"stockholm\", etc), which is used to turn the alignment into a\n",
    "        string.\n",
    "\n",
    "        e.g.\n",
    "\n",
    "        >>> from Bio.Alphabet import IUPAC, Gapped\n",
    "        >>> from Bio.Align import MultipleSeqAlignment\n",
    "        >>> align = MultipleSeqAlignment([], Gapped(IUPAC.unambiguous_dna, \"-\"))\n",
    "        >>> align.add_sequence(\"Alpha\", \"ACTGCTAGCTAG\")\n",
    "        >>> align.add_sequence(\"Beta\",  \"ACT-CTAGCTAG\")\n",
    "        >>> align.add_sequence(\"Gamma\", \"ACTGCTAGATAG\")\n",
    "        >>> print(align.format(\"fasta\"))\n",
    "        >Alpha\n",
    "        ACTGCTAGCTAG\n",
    "        >Beta\n",
    "        ACT-CTAGCTAG\n",
    "        >Gamma\n",
    "        ACTGCTAGATAG\n",
    "        <BLANKLINE>\n",
    "        >>> print(align.format(\"phylip\"))\n",
    "         3 12\n",
    "        Alpha      ACTGCTAGCT AG\n",
    "        Beta       ACT-CTAGCT AG\n",
    "        Gamma      ACTGCTAGAT AG\n",
    "        <BLANKLINE>\n",
    "\n",
    "        For Python 2.6, 3.0 or later see also the built in format() function.\n",
    "        \"\"\"\n",
    "        # See also the __format__ added for Python 2.6 / 3.0, PEP 3101\n",
    "        # See also the SeqRecord class and its format() method using Bio.SeqIO\n",
    "        return self.__format__(format)\n",
    "\n",
    "    def __format__(self, format_spec):\n",
    "        \"\"\"Return the alignment as a string in the specified file format.\n",
    "\n",
    "        This method supports the python format() function added in\n",
    "        Python 2.6/3.0.  The format_spec should be a lower case\n",
    "        string supported by Bio.AlignIO as an output file format.\n",
    "        See also the alignment's format() method.\n",
    "        \"\"\"\n",
    "        if format_spec:\n",
    "            from Bio._py3k import StringIO\n",
    "            from Bio import AlignIO\n",
    "            handle = StringIO()\n",
    "            AlignIO.write([self], handle, format_spec)\n",
    "            return handle.getvalue()\n",
    "        else:\n",
    "            # Follow python convention and default to using __str__\n",
    "            return str(self)\n",
    "\n",
    "    def __iter__(self):\n",
    "        \"\"\"Iterate over alignment rows as SeqRecord objects.\n",
    "\n",
    "        e.g.\n",
    "\n",
    "        >>> from Bio.Alphabet import IUPAC, Gapped\n",
    "        >>> from Bio.Align import MultipleSeqAlignment\n",
    "        >>> align = MultipleSeqAlignment([], Gapped(IUPAC.unambiguous_dna, \"-\"))\n",
    "        >>> align.add_sequence(\"Alpha\", \"ACTGCTAGCTAG\")\n",
    "        >>> align.add_sequence(\"Beta\",  \"ACT-CTAGCTAG\")\n",
    "        >>> align.add_sequence(\"Gamma\", \"ACTGCTAGATAG\")\n",
    "        >>> for record in align:\n",
    "        ...    print(record.id)\n",
    "        ...    print(record.seq)\n",
    "        ...\n",
    "        Alpha\n",
    "        ACTGCTAGCTAG\n",
    "        Beta\n",
    "        ACT-CTAGCTAG\n",
    "        Gamma\n",
    "        ACTGCTAGATAG\n",
    "        \"\"\"\n",
    "        return iter(self._records)\n",
    "\n",
    "    def __len__(self):\n",
    "        \"\"\"Return the number of sequences in the alignment.\n",
    "\n",
    "        Use len(alignment) to get the number of sequences (i.e. the number of\n",
    "        rows), and alignment.get_alignment_length() to get the length of the\n",
    "        longest sequence (i.e. the number of columns).\n",
    "\n",
    "        This is easy to remember if you think of the alignment as being like a\n",
    "        list of SeqRecord objects.\n",
    "        \"\"\"\n",
    "        return len(self._records)\n",
    "\n",
    "    def get_alignment_length(self):\n",
    "        \"\"\"Return the maximum length of the alignment.\n",
    "\n",
    "        All objects in the alignment should (hopefully) have the same\n",
    "        length. This function will go through and find this length\n",
    "        by finding the maximum length of sequences in the alignment.\n",
    "\n",
    "        >>> from Bio.Alphabet import IUPAC, Gapped\n",
    "        >>> from Bio.Align import MultipleSeqAlignment\n",
    "        >>> align = MultipleSeqAlignment([], Gapped(IUPAC.unambiguous_dna, \"-\"))\n",
    "        >>> align.add_sequence(\"Alpha\", \"ACTGCTAGCTAG\")\n",
    "        >>> align.add_sequence(\"Beta\",  \"ACT-CTAGCTAG\")\n",
    "        >>> align.add_sequence(\"Gamma\", \"ACTGCTAGATAG\")\n",
    "        >>> align.get_alignment_length()\n",
    "        12\n",
    "\n",
    "        If you want to know the number of sequences in the alignment,\n",
    "        use len(align) instead:\n",
    "\n",
    "        >>> len(align)\n",
    "        3\n",
    "\n",
    "        \"\"\"\n",
    "        max_length = 0\n",
    "\n",
    "        for record in self._records:\n",
    "            if len(record.seq) > max_length:\n",
    "                max_length = len(record.seq)\n",
    "\n",
    "        return max_length\n",
    "\n",
    "    def add_sequence(self, descriptor, sequence, start=None, end=None,\n",
    "                     weight=1.0):\n",
    "        \"\"\"Add a sequence to the alignment.\n",
    "\n",
    "        This doesn't do any kind of alignment, it just adds in the sequence\n",
    "        object, which is assumed to be prealigned with the existing\n",
    "        sequences.\n",
    "\n",
    "        Arguments:\n",
    "            - descriptor - The descriptive id of the sequence being added.\n",
    "              This will be used as the resulting SeqRecord's\n",
    "              .id property (and, for historical compatibility,\n",
    "              also the .description property)\n",
    "            - sequence - A string with sequence info.\n",
    "            - start - You can explicitly set the start point of the sequence.\n",
    "              This is useful (at least) for BLAST alignments, which can\n",
    "              just be partial alignments of sequences.\n",
    "            - end - Specify the end of the sequence, which is important\n",
    "              for the same reason as the start.\n",
    "            - weight - The weight to place on the sequence in the alignment.\n",
    "              By default, all sequences have the same weight. (0.0 =>\n",
    "              no weight, 1.0 => highest weight)\n",
    "\n",
    "        In general providing a SeqRecord and calling .append is preferred.\n",
    "        \"\"\"\n",
    "        new_seq = Seq(sequence, self._alphabet)\n",
    "\n",
    "        # We are now effectively using the SeqRecord's .id as\n",
    "        # the primary identifier (e.g. in Bio.SeqIO) so we should\n",
    "        # populate it with the descriptor.\n",
    "        # For backwards compatibility, also store this in the\n",
    "        # SeqRecord's description property.\n",
    "        new_record = SeqRecord(new_seq,\n",
    "                               id=descriptor,\n",
    "                               description=descriptor)\n",
    "\n",
    "        # hack! We really need to work out how to deal with annotations\n",
    "        # and features in biopython. Right now, I'll just use the\n",
    "        # generic annotations dictionary we've got to store the start\n",
    "        # and end, but we should think up something better. I don't know\n",
    "        # if I'm really a big fan of the LocatableSeq thing they've got\n",
    "        # in BioPerl, but I'm not positive what the best thing to do on\n",
    "        # this is...\n",
    "        if start:\n",
    "            new_record.annotations[\"start\"] = start\n",
    "        if end:\n",
    "            new_record.annotations[\"end\"] = end\n",
    "\n",
    "        # another hack to add weight information to the sequence\n",
    "        new_record.annotations[\"weight\"] = weight\n",
    "\n",
    "        self._records.append(new_record)\n",
    "\n",
    "    def extend(self, records):\n",
    "        \"\"\"Add more SeqRecord objects to the alignment as rows.\n",
    "\n",
    "        They must all have the same length as the original alignment, and have\n",
    "        alphabets compatible with the alignment's alphabet. For example,\n",
    "\n",
    "        >>> from Bio.Alphabet import generic_dna\n",
    "        >>> from Bio.Seq import Seq\n",
    "        >>> from Bio.SeqRecord import SeqRecord\n",
    "        >>> from Bio.Align import MultipleSeqAlignment\n",
    "        >>> a = SeqRecord(Seq(\"AAAACGT\", generic_dna), id=\"Alpha\")\n",
    "        >>> b = SeqRecord(Seq(\"AAA-CGT\", generic_dna), id=\"Beta\")\n",
    "        >>> c = SeqRecord(Seq(\"AAAAGGT\", generic_dna), id=\"Gamma\")\n",
    "        >>> d = SeqRecord(Seq(\"AAAACGT\", generic_dna), id=\"Delta\")\n",
    "        >>> e = SeqRecord(Seq(\"AAA-GGT\", generic_dna), id=\"Epsilon\")\n",
    "\n",
    "        First we create a small alignment (three rows):\n",
    "\n",
    "        >>> align = MultipleSeqAlignment([a, b, c])\n",
    "        >>> print(align)\n",
    "        DNAAlphabet() alignment with 3 rows and 7 columns\n",
    "        AAAACGT Alpha\n",
    "        AAA-CGT Beta\n",
    "        AAAAGGT Gamma\n",
    "\n",
    "        Now we can extend this alignment with another two rows:\n",
    "\n",
    "        >>> align.extend([d, e])\n",
    "        >>> print(align)\n",
    "        DNAAlphabet() alignment with 5 rows and 7 columns\n",
    "        AAAACGT Alpha\n",
    "        AAA-CGT Beta\n",
    "        AAAAGGT Gamma\n",
    "        AAAACGT Delta\n",
    "        AAA-GGT Epsilon\n",
    "\n",
    "        Because the alignment object allows iteration over the rows as\n",
    "        SeqRecords, you can use the extend method with a second alignment\n",
    "        (provided its sequences have the same length as the original alignment).\n",
    "        \"\"\"\n",
    "        if len(self):\n",
    "            # Use the standard method to get the length\n",
    "            expected_length = self.get_alignment_length()\n",
    "        else:\n",
    "            # Take the first record's length\n",
    "            records = iter(records)  # records arg could be list or iterator\n",
    "            try:\n",
    "                rec = next(records)\n",
    "            except StopIteration:\n",
    "                # Special case, no records\n",
    "                return\n",
    "            expected_length = len(rec)\n",
    "            self._append(rec, expected_length)\n",
    "            # Can now setup the per-column-annotations as well, set to None\n",
    "            # while missing the length:\n",
    "            self.column_annotations = {}\n",
    "            # Now continue to the rest of the records as usual\n",
    "\n",
    "        for rec in records:\n",
    "            self._append(rec, expected_length)\n",
    "\n",
    "    def append(self, record):\n",
    "        \"\"\"Add one more SeqRecord object to the alignment as a new row.\n",
    "\n",
    "        This must have the same length as the original alignment (unless this is\n",
    "        the first record), and have an alphabet compatible with the alignment's\n",
    "        alphabet.\n",
    "\n",
    "        >>> from Bio import AlignIO\n",
    "        >>> align = AlignIO.read(\"Clustalw/opuntia.aln\", \"clustal\")\n",
    "        >>> print(align)\n",
    "        SingleLetterAlphabet() alignment with 7 rows and 156 columns\n",
    "        TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191\n",
    "        TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191\n",
    "        TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191\n",
    "        TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191\n",
    "        TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191\n",
    "        TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191\n",
    "        TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191\n",
    "        >>> len(align)\n",
    "        7\n",
    "\n",
    "        We'll now construct a dummy record to append as an example:\n",
    "\n",
    "        >>> from Bio.Seq import Seq\n",
    "        >>> from Bio.SeqRecord import SeqRecord\n",
    "        >>> dummy = SeqRecord(Seq(\"N\"*156), id=\"dummy\")\n",
    "\n",
    "        Now append this to the alignment,\n",
    "\n",
    "        >>> align.append(dummy)\n",
    "        >>> print(align)\n",
    "        SingleLetterAlphabet() alignment with 8 rows and 156 columns\n",
    "        TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191\n",
    "        TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191\n",
    "        TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191\n",
    "        TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191\n",
    "        TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191\n",
    "        TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191\n",
    "        TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191\n",
    "        NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN dummy\n",
    "        >>> len(align)\n",
    "        8\n",
    "\n",
    "        \"\"\"\n",
    "        if self._records:\n",
    "            self._append(record, self.get_alignment_length())\n",
    "        else:\n",
    "            self._append(record)\n",
    "\n",
    "    def _append(self, record, expected_length=None):\n",
    "        \"\"\"Validate and append a record (PRIVATE).\"\"\"\n",
    "        if not isinstance(record, SeqRecord):\n",
    "            raise TypeError(\"New sequence is not a SeqRecord object\")\n",
    "\n",
    "        # Currently the get_alignment_length() call is expensive, so we need\n",
    "        # to avoid calling it repeatedly for __init__ and extend, hence this\n",
    "        # private _append method\n",
    "        if expected_length is not None and len(record) != expected_length:\n",
    "            # TODO - Use the following more helpful error, but update unit tests\n",
    "            # raise ValueError(\"New sequence is not of length %i\" \\\n",
    "            #                 % self.get_alignment_length())\n",
    "            raise ValueError(\"Sequences must all be the same length\")\n",
    "\n",
    "        # Using not self.alphabet.contains(record.seq.alphabet) needs fixing\n",
    "        # for AlphabetEncoders (e.g. gapped versus ungapped).\n",
    "        if not Alphabet._check_type_compatible([self._alphabet, record.seq.alphabet]):\n",
    "            raise ValueError(\"New sequence's alphabet is incompatible\")\n",
    "        self._records.append(record)\n",
    "\n",
    "    def __add__(self, other):\n",
    "        \"\"\"Combine two alignments with the same number of rows by adding them.\n",
    "\n",
    "        If you have two multiple sequence alignments (MSAs), there are two ways to think\n",
    "        about adding them - by row or by column. Using the extend method adds by row.\n",
    "        Using the addition operator adds by column. For example,\n",
    "\n",
    "        >>> from Bio.Alphabet import generic_dna\n",
    "        >>> from Bio.Seq import Seq\n",
    "        >>> from Bio.SeqRecord import SeqRecord\n",
    "        >>> from Bio.Align import MultipleSeqAlignment\n",
    "        >>> a1 = SeqRecord(Seq(\"AAAAC\", generic_dna), id=\"Alpha\")\n",
    "        >>> b1 = SeqRecord(Seq(\"AAA-C\", generic_dna), id=\"Beta\")\n",
    "        >>> c1 = SeqRecord(Seq(\"AAAAG\", generic_dna), id=\"Gamma\")\n",
    "        >>> a2 = SeqRecord(Seq(\"GT\", generic_dna), id=\"Alpha\")\n",
    "        >>> b2 = SeqRecord(Seq(\"GT\", generic_dna), id=\"Beta\")\n",
    "        >>> c2 = SeqRecord(Seq(\"GT\", generic_dna), id=\"Gamma\")\n",
    "        >>> left = MultipleSeqAlignment([a1, b1, c1],\n",
    "        ...                             annotations={\"tool\": \"demo\", \"name\": \"start\"},\n",
    "        ...                             column_annotations={\"stats\": \"CCCXC\"})\n",
    "        >>> right = MultipleSeqAlignment([a2, b2, c2],\n",
    "        ...                             annotations={\"tool\": \"demo\", \"name\": \"end\"},\n",
    "        ...                             column_annotations={\"stats\": \"CC\"})\n",
    "\n",
    "        Now, let's look at these two alignments:\n",
    "\n",
    "        >>> print(left)\n",
    "        DNAAlphabet() alignment with 3 rows and 5 columns\n",
    "        AAAAC Alpha\n",
    "        AAA-C Beta\n",
    "        AAAAG Gamma\n",
    "        >>> print(right)\n",
    "        DNAAlphabet() alignment with 3 rows and 2 columns\n",
    "        GT Alpha\n",
    "        GT Beta\n",
    "        GT Gamma\n",
    "\n",
    "        And add them:\n",
    "\n",
    "        >>> combined = left + right\n",
    "        >>> print(combined)\n",
    "        DNAAlphabet() alignment with 3 rows and 7 columns\n",
    "        AAAACGT Alpha\n",
    "        AAA-CGT Beta\n",
    "        AAAAGGT Gamma\n",
    "\n",
    "        For this to work, both alignments must have the same number of records (here\n",
    "        they both have 3 rows):\n",
    "\n",
    "        >>> len(left)\n",
    "        3\n",
    "        >>> len(right)\n",
    "        3\n",
    "        >>> len(combined)\n",
    "        3\n",
    "\n",
    "        The individual rows are SeqRecord objects, and these can be added together. Refer\n",
    "        to the SeqRecord documentation for details of how the annotation is handled. This\n",
    "        example is a special case in that both original alignments shared the same names,\n",
    "        meaning when the rows are added they also get the same name.\n",
    "\n",
    "        Any common annotations are preserved, but differing annotation is lost. This is\n",
    "        the same behaviour used in the SeqRecord annotations and is designed to prevent\n",
    "        accidental propagation of inappropriate values:\n",
    "\n",
    "        >>> combined.annotations\n",
    "        {'tool': 'demo'}\n",
    "\n",
    "        Similarly any common per-column-annotations are combined:\n",
    "\n",
    "        >>> combined.column_annotations\n",
    "        {'stats': 'CCCXCCC'}\n",
    "\n",
    "        \"\"\"\n",
    "        if not isinstance(other, MultipleSeqAlignment):\n",
    "            raise NotImplementedError\n",
    "        if len(self) != len(other):\n",
    "            raise ValueError(\"When adding two alignments they must have the same length\"\n",
    "                             \" (i.e. same number or rows)\")\n",
    "        alpha = Alphabet._consensus_alphabet([self._alphabet, other._alphabet])\n",
    "        merged = (left + right for left, right in zip(self, other))\n",
    "        # Take any common annotation:\n",
    "        annotations = {}\n",
    "        for k, v in self.annotations.items():\n",
    "            if k in other.annotations and other.annotations[k] == v:\n",
    "                annotations[k] = v\n",
    "        column_annotations = {}\n",
    "        for k, v in self.column_annotations.items():\n",
    "            if k in other.column_annotations:\n",
    "                column_annotations[k] = v + other.column_annotations[k]\n",
    "        return MultipleSeqAlignment(merged, alpha, annotations, column_annotations)\n",
    "\n",
    "    def __getitem__(self, index):\n",
    "        \"\"\"Access part of the alignment.\n",
    "\n",
    "        Depending on the indices, you can get a SeqRecord object\n",
    "        (representing a single row), a Seq object (for a single columns),\n",
    "        a string (for a single characters) or another alignment\n",
    "        (representing some part or all of the alignment).\n",
    "\n",
    "        align[r,c] gives a single character as a string\n",
    "        align[r] gives a row as a SeqRecord\n",
    "        align[r,:] gives a row as a SeqRecord\n",
    "        align[:,c] gives a column as a Seq (using the alignment's alphabet)\n",
    "\n",
    "        align[:] and align[:,:] give a copy of the alignment\n",
    "\n",
    "        Anything else gives a sub alignment, e.g.\n",
    "        align[0:2] or align[0:2,:] uses only row 0 and 1\n",
    "        align[:,1:3] uses only columns 1 and 2\n",
    "        align[0:2,1:3] uses only rows 0 & 1 and only cols 1 & 2\n",
    "\n",
    "        We'll use the following example alignment here for illustration:\n",
    "\n",
    "        >>> from Bio.Alphabet import generic_dna\n",
    "        >>> from Bio.Seq import Seq\n",
    "        >>> from Bio.SeqRecord import SeqRecord\n",
    "        >>> from Bio.Align import MultipleSeqAlignment\n",
    "        >>> a = SeqRecord(Seq(\"AAAACGT\", generic_dna), id=\"Alpha\")\n",
    "        >>> b = SeqRecord(Seq(\"AAA-CGT\", generic_dna), id=\"Beta\")\n",
    "        >>> c = SeqRecord(Seq(\"AAAAGGT\", generic_dna), id=\"Gamma\")\n",
    "        >>> d = SeqRecord(Seq(\"AAAACGT\", generic_dna), id=\"Delta\")\n",
    "        >>> e = SeqRecord(Seq(\"AAA-GGT\", generic_dna), id=\"Epsilon\")\n",
    "        >>> align = MultipleSeqAlignment([a, b, c, d, e], generic_dna)\n",
    "\n",
    "        You can access a row of the alignment as a SeqRecord using an integer\n",
    "        index (think of the alignment as a list of SeqRecord objects here):\n",
    "\n",
    "        >>> first_record = align[0]\n",
    "        >>> print(\"%s %s\" % (first_record.id, first_record.seq))\n",
    "        Alpha AAAACGT\n",
    "        >>> last_record = align[-1]\n",
    "        >>> print(\"%s %s\" % (last_record.id, last_record.seq))\n",
    "        Epsilon AAA-GGT\n",
    "\n",
    "        You can also access use python's slice notation to create a sub-alignment\n",
    "        containing only some of the SeqRecord objects:\n",
    "\n",
    "        >>> sub_alignment = align[2:5]\n",
    "        >>> print(sub_alignment)\n",
    "        DNAAlphabet() alignment with 3 rows and 7 columns\n",
    "        AAAAGGT Gamma\n",
    "        AAAACGT Delta\n",
    "        AAA-GGT Epsilon\n",
    "\n",
    "        This includes support for a step, i.e. align[start:end:step], which\n",
    "        can be used to select every second sequence:\n",
    "\n",
    "        >>> sub_alignment = align[::2]\n",
    "        >>> print(sub_alignment)\n",
    "        DNAAlphabet() alignment with 3 rows and 7 columns\n",
    "        AAAACGT Alpha\n",
    "        AAAAGGT Gamma\n",
    "        AAA-GGT Epsilon\n",
    "\n",
    "        Or to get a copy of the alignment with the rows in reverse order:\n",
    "\n",
    "        >>> rev_alignment = align[::-1]\n",
    "        >>> print(rev_alignment)\n",
    "        DNAAlphabet() alignment with 5 rows and 7 columns\n",
    "        AAA-GGT Epsilon\n",
    "        AAAACGT Delta\n",
    "        AAAAGGT Gamma\n",
    "        AAA-CGT Beta\n",
    "        AAAACGT Alpha\n",
    "\n",
    "        You can also use two indices to specify both rows and columns. Using simple\n",
    "        integers gives you the entry as a single character string. e.g.\n",
    "\n",
    "        >>> align[3, 4]\n",
    "        'C'\n",
    "\n",
    "        This is equivalent to:\n",
    "\n",
    "        >>> align[3][4]\n",
    "        'C'\n",
    "\n",
    "        or:\n",
    "\n",
    "        >>> align[3].seq[4]\n",
    "        'C'\n",
    "\n",
    "        To get a single column (as a string) use this syntax:\n",
    "\n",
    "        >>> align[:, 4]\n",
    "        'CCGCG'\n",
    "\n",
    "        Or, to get part of a column,\n",
    "\n",
    "        >>> align[1:3, 4]\n",
    "        'CG'\n",
    "\n",
    "        However, in general you get a sub-alignment,\n",
    "\n",
    "        >>> print(align[1:5, 3:6])\n",
    "        DNAAlphabet() alignment with 4 rows and 3 columns\n",
    "        -CG Beta\n",
    "        AGG Gamma\n",
    "        ACG Delta\n",
    "        -GG Epsilon\n",
    "\n",
    "        This should all seem familiar to anyone who has used the NumPy\n",
    "        array or matrix objects.\n",
    "        \"\"\"\n",
    "        if isinstance(index, int):\n",
    "            # e.g. result = align[x]\n",
    "            # Return a SeqRecord\n",
    "            return self._records[index]\n",
    "        elif isinstance(index, slice):\n",
    "            # e.g. sub_align = align[i:j:k]\n",
    "            new = MultipleSeqAlignment(self._records[index], self._alphabet)\n",
    "            if self.column_annotations and len(new) == len(self):\n",
    "                # All rows kept (although could have been reversed)\n",
    "                # Perserve the column annotations too,\n",
    "                for k, v in self.column_annotations.items():\n",
    "                    new.column_annotations[k] = v\n",
    "            return new\n",
    "        elif len(index) != 2:\n",
    "            raise TypeError(\"Invalid index type.\")\n",
    "\n",
    "        # Handle double indexing\n",
    "        row_index, col_index = index\n",
    "        if isinstance(row_index, int):\n",
    "            # e.g. row_or_part_row = align[6, 1:4], gives a SeqRecord\n",
    "            return self._records[row_index][col_index]\n",
    "        elif isinstance(col_index, int):\n",
    "            # e.g. col_or_part_col = align[1:5, 6], gives a string\n",
    "            return \"\".join(rec[col_index] for rec in self._records[row_index])\n",
    "        else:\n",
    "            # e.g. sub_align = align[1:4, 5:7], gives another alignment\n",
    "            new = MultipleSeqAlignment((rec[col_index] for rec in self._records[row_index]),\n",
    "                                       self._alphabet)\n",
    "            if self.column_annotations and len(new) == len(self):\n",
    "                # All rows kept (although could have been reversed)\n",
    "                # Perserve the column annotations too,\n",
    "                for k, v in self.column_annotations.items():\n",
    "                    new.column_annotations[k] = v[col_index]\n",
    "            return new\n",
    "\n",
    "    def sort(self, key=None, reverse=False):\n",
    "        \"\"\"Sort the rows (SeqRecord objects) of the alignment in place.\n",
    "\n",
    "        This sorts the rows alphabetically using the SeqRecord object id by\n",
    "        default. The sorting can be controlled by supplying a key function\n",
    "        which must map each SeqRecord to a sort value.\n",
    "\n",
    "        This is useful if you want to add two alignments which use the same\n",
    "        record identifiers, but in a different order. For example,\n",
    "\n",
    "        >>> from Bio.Alphabet import generic_dna\n",
    "        >>> from Bio.Seq import Seq\n",
    "        >>> from Bio.SeqRecord import SeqRecord\n",
    "        >>> from Bio.Align import MultipleSeqAlignment\n",
    "        >>> align1 = MultipleSeqAlignment([\n",
    "        ...              SeqRecord(Seq(\"ACGT\", generic_dna), id=\"Human\"),\n",
    "        ...              SeqRecord(Seq(\"ACGG\", generic_dna), id=\"Mouse\"),\n",
    "        ...              SeqRecord(Seq(\"ACGC\", generic_dna), id=\"Chicken\"),\n",
    "        ...          ])\n",
    "        >>> align2 = MultipleSeqAlignment([\n",
    "        ...              SeqRecord(Seq(\"CGGT\", generic_dna), id=\"Mouse\"),\n",
    "        ...              SeqRecord(Seq(\"CGTT\", generic_dna), id=\"Human\"),\n",
    "        ...              SeqRecord(Seq(\"CGCT\", generic_dna), id=\"Chicken\"),\n",
    "        ...          ])\n",
    "\n",
    "        If you simple try and add these without sorting, you get this:\n",
    "\n",
    "        >>> print(align1 + align2)\n",
    "        DNAAlphabet() alignment with 3 rows and 8 columns\n",
    "        ACGTCGGT <unknown id>\n",
    "        ACGGCGTT <unknown id>\n",
    "        ACGCCGCT Chicken\n",
    "\n",
    "        Consult the SeqRecord documentation which explains why you get a\n",
    "        default value when annotation like the identifier doesn't match up.\n",
    "        However, if we sort the alignments first, then add them we get the\n",
    "        desired result:\n",
    "\n",
    "        >>> align1.sort()\n",
    "        >>> align2.sort()\n",
    "        >>> print(align1 + align2)\n",
    "        DNAAlphabet() alignment with 3 rows and 8 columns\n",
    "        ACGCCGCT Chicken\n",
    "        ACGTCGTT Human\n",
    "        ACGGCGGT Mouse\n",
    "\n",
    "        As an example using a different sort order, you could sort on the\n",
    "        GC content of each sequence.\n",
    "\n",
    "        >>> from Bio.SeqUtils import GC\n",
    "        >>> print(align1)\n",
    "        DNAAlphabet() alignment with 3 rows and 4 columns\n",
    "        ACGC Chicken\n",
    "        ACGT Human\n",
    "        ACGG Mouse\n",
    "        >>> align1.sort(key = lambda record: GC(record.seq))\n",
    "        >>> print(align1)\n",
    "        DNAAlphabet() alignment with 3 rows and 4 columns\n",
    "        ACGT Human\n",
    "        ACGC Chicken\n",
    "        ACGG Mouse\n",
    "\n",
    "        There is also a reverse argument, so if you wanted to sort by ID\n",
    "        but backwards:\n",
    "\n",
    "        >>> align1.sort(reverse=True)\n",
    "        >>> print(align1)\n",
    "        DNAAlphabet() alignment with 3 rows and 4 columns\n",
    "        ACGG Mouse\n",
    "        ACGT Human\n",
    "        ACGC Chicken\n",
    "\n",
    "        \"\"\"\n",
    "        if key is None:\n",
    "            self._records.sort(key=lambda r: r.id, reverse=reverse)\n",
    "        else:\n",
    "            self._records.sort(key=key, reverse=reverse)\n",
    "\n",
    "\n",
    "class PairwiseAlignment(object):\n",
    "    \"\"\"Represents a pairwise sequence alignment.\n",
    "\n",
    "    Internally, the pairwise alignment is stored as the path through\n",
    "    the traceback matrix, i.e. a tuple of pairs of indices corresponding\n",
    "    to the vertices of the path in the traceback matrix.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, target, query, path, score):\n",
    "        \"\"\"Initialize a new PairwiseAlignment object.\n",
    "\n",
    "        Arguments:\n",
    "         - target  - The first sequence, as a plain string, without gaps.\n",
    "         - query   - The second sequence, as a plain string, without gaps.\n",
    "         - path    - The path through the traceback matrix, defining an\n",
    "                     alignment.\n",
    "         - score   - The alignment score.\n",
    "\n",
    "        You would normally obtain a PairwiseAlignment object by iterating\n",
    "        over a PairwiseAlignments object.\n",
    "        \"\"\"\n",
    "        self.target = target\n",
    "        self.query = query\n",
    "        self.score = score\n",
    "        self.path = path\n",
    "\n",
    "    # For Python2 only\n",
    "    def __cmp__(self, other):\n",
    "        if self.path < other.path:\n",
    "            return -1\n",
    "        if self.path > other.path:\n",
    "            return +1\n",
    "        return 0\n",
    "\n",
    "    def __eq__(self, other):\n",
    "        return self.path == other.path\n",
    "\n",
    "    def __ne__(self, other):\n",
    "        return self.path != other.path\n",
    "\n",
    "    def __lt__(self, other):\n",
    "        return self.path < other.path\n",
    "\n",
    "    def __le__(self, other):\n",
    "        return self.path <= other.path\n",
    "\n",
    "    def __gt__(self, other):\n",
    "        return self.path > other.path\n",
    "\n",
    "    def __ge__(self, other):\n",
    "        return self.path >= other.path\n",
    "\n",
    "    def __format__(self, format_spec):\n",
    "        if format_spec == \"psl\":\n",
    "            return self._format_psl()\n",
    "        return str(self)\n",
    "\n",
    "    def __str__(self):\n",
    "        if isinstance(self.query, str) and isinstance(self.target, str):\n",
    "            return self.format()\n",
    "        else:\n",
    "            return self._format_generalized()\n",
    "\n",
    "    def format(self):\n",
    "        \"\"\"Create a human-readable representation of the alignment.\"\"\"\n",
    "        query = self.query\n",
    "        target = self.target\n",
    "        try:\n",
    "            # check if query is a SeqRecord\n",
    "            query = query.seq\n",
    "        except AttributeError:\n",
    "            # query is a Seq object or a plain string\n",
    "            pass\n",
    "        try:\n",
    "            # check if target is a SeqRecord\n",
    "            target = target.seq\n",
    "        except AttributeError:\n",
    "            # target is a Seq object or a plain string\n",
    "            pass\n",
    "        seq1 = str(target)\n",
    "        seq2 = str(query)\n",
    "        n1 = len(seq1)\n",
    "        n2 = len(seq2)\n",
    "        aligned_seq1 = \"\"\n",
    "        aligned_seq2 = \"\"\n",
    "        pattern = \"\"\n",
    "        path = self.path\n",
    "        end1, end2 = path[0]\n",
    "        if end1 > 0 or end2 > 0:\n",
    "            end = max(end1, end2)\n",
    "            aligned_seq1 += \" \" * (end - end1) + seq1[:end1]\n",
    "            aligned_seq2 += \" \" * (end - end2) + seq2[:end2]\n",
    "            pattern += \" \" * end\n",
    "        start1 = end1\n",
    "        start2 = end2\n",
    "        for end1, end2 in path[1:]:\n",
    "            gap = 0\n",
    "            if end1 == start1:\n",
    "                gap = end2 - start2\n",
    "                aligned_seq1 += \"-\" * gap\n",
    "                aligned_seq2 += seq2[start2:end2]\n",
    "                pattern += \"-\" * gap\n",
    "            elif end2 == start2:\n",
    "                gap = end1 - start1\n",
    "                aligned_seq1 += seq1[start1:end1]\n",
    "                aligned_seq2 += \"-\" * gap\n",
    "                pattern += \"-\" * gap\n",
    "            else:\n",
    "                s1 = seq1[start1:end1]\n",
    "                s2 = seq2[start2:end2]\n",
    "                aligned_seq1 += s1\n",
    "                aligned_seq2 += s2\n",
    "                for c1, c2 in zip(s1, s2):\n",
    "                    if c1 == c2:\n",
    "                        pattern += \"|\"\n",
    "                    else:\n",
    "                        pattern += \".\"\n",
    "            start1 = end1\n",
    "            start2 = end2\n",
    "        n1 -= end1\n",
    "        n2 -= end2\n",
    "        n = max(n1, n2)\n",
    "        aligned_seq1 += seq1[end1:] + \" \" * (n - n1)\n",
    "        aligned_seq2 += seq2[end2:] + \" \" * (n - n2)\n",
    "        pattern += \" \" * n\n",
    "        return \"%s\\n%s\\n%s\\n\" % (aligned_seq1, pattern, aligned_seq2)\n",
    "\n",
    "    def _format_generalized(self):\n",
    "        seq1 = self.target\n",
    "        seq2 = self.query\n",
    "        n1 = len(seq1)\n",
    "        n2 = len(seq2)\n",
    "        aligned_seq1 = []\n",
    "        aligned_seq2 = []\n",
    "        pattern = []\n",
    "        path = self.path\n",
    "        end1, end2 = path[0]\n",
    "        if end1 > 0 or end2 > 0:\n",
    "            if end1 <= end2:\n",
    "                for c2 in seq2[:end2 - end1]:\n",
    "                    s2 = str(c2)\n",
    "                    s1 = \" \" * len(s2)\n",
    "                    aligned_seq1.append(s1)\n",
    "                    aligned_seq2.append(s2)\n",
    "                    pattern.append(s1)\n",
    "            else:  # end1 > end2\n",
    "                for c1 in seq1[:end1 - end2]:\n",
    "                    s1 = str(c1)\n",
    "                    s2 = \" \" * len(s1)\n",
    "                    aligned_seq1.append(s1)\n",
    "                    aligned_seq2.append(s2)\n",
    "                    pattern.append(s2)\n",
    "        start1 = end1\n",
    "        start2 = end2\n",
    "        for end1, end2 in path[1:]:\n",
    "            if end1 == start1:\n",
    "                for c2 in seq2[start2:end2]:\n",
    "                    s2 = str(c2)\n",
    "                    s1 = \"-\" * len(s2)\n",
    "                    aligned_seq1.append(s1)\n",
    "                    aligned_seq2.append(s2)\n",
    "                    pattern.append(s1)\n",
    "                start2 = end2\n",
    "            elif end2 == start2:\n",
    "                for c1 in seq1[start1:end1]:\n",
    "                    s1 = str(c1)\n",
    "                    s2 = \"-\" * len(s1)\n",
    "                    aligned_seq1.append(s1)\n",
    "                    aligned_seq2.append(s2)\n",
    "                    pattern.append(s2)\n",
    "                start1 = end1\n",
    "            else:\n",
    "                for c1, c2 in zip(seq1[start1:end1], seq2[start2:end2]):\n",
    "                    s1 = str(c1)\n",
    "                    s2 = str(c2)\n",
    "                    m1 = len(s1)\n",
    "                    m2 = len(s2)\n",
    "                    if c1 == c2:\n",
    "                        p = \"|\"\n",
    "                    else:\n",
    "                        p = \".\"\n",
    "                    if m1 < m2:\n",
    "                        space = (m2 - m1) * \" \"\n",
    "                        s1 += space\n",
    "                        pattern.append(p * m1 + space)\n",
    "                    elif m1 > m2:\n",
    "                        space = (m1 - m2) * \" \"\n",
    "                        s2 += space\n",
    "                        pattern.append(p * m2 + space)\n",
    "                    else:\n",
    "                        pattern.append(p * m1)\n",
    "                    aligned_seq1.append(s1)\n",
    "                    aligned_seq2.append(s2)\n",
    "                start1 = end1\n",
    "                start2 = end2\n",
    "        aligned_seq1 = \" \".join(aligned_seq1)\n",
    "        aligned_seq2 = \" \".join(aligned_seq2)\n",
    "        pattern = \" \".join(pattern)\n",
    "        return \"%s\\n%s\\n%s\\n\" % (aligned_seq1, pattern, aligned_seq2)\n",
    "\n",
    "    def _format_psl(self):\n",
    "        query = self.query\n",
    "        target = self.target\n",
    "        try:\n",
    "            Qname = query.id\n",
    "        except AttributeError:\n",
    "            Qname = \"query\"\n",
    "        else:\n",
    "            query = query.seq\n",
    "        try:\n",
    "            Tname = target.id\n",
    "        except AttributeError:\n",
    "            Tname = \"target\"\n",
    "        else:\n",
    "            target = target.seq\n",
    "        seq1 = str(target)\n",
    "        seq2 = str(query)\n",
    "        n1 = len(seq1)\n",
    "        n2 = len(seq2)\n",
    "        match = 0\n",
    "        mismatch = 0\n",
    "        repmatch = 0\n",
    "        Ns = 0\n",
    "        Qgapcount = 0\n",
    "        Qgapbases = 0\n",
    "        Tgapcount = 0\n",
    "        Tgapbases = 0\n",
    "        Qsize = n2\n",
    "        Qstart = 0\n",
    "        Qend = Qsize\n",
    "        Tsize = n1\n",
    "        Tstart = 0\n",
    "        Tend = Tsize\n",
    "        blockSizes = []\n",
    "        qStarts = []\n",
    "        tStarts = []\n",
    "        strand = \"+\"\n",
    "        start1 = 0\n",
    "        start2 = 0\n",
    "        start1, start2 = self.path[0]\n",
    "        for end1, end2 in self.path[1:]:\n",
    "            count1 = end1 - start1\n",
    "            count2 = end2 - start2\n",
    "            if count1 == 0:\n",
    "                if start2 == 0:\n",
    "                    Qstart += count2\n",
    "                elif end2 == n2:\n",
    "                    Qend -= count2\n",
    "                else:\n",
    "                    Qgapcount += 1\n",
    "                    Qgapbases += count2\n",
    "                start2 = end2\n",
    "            elif count2 == 0:\n",
    "                if start1 == 0:\n",
    "                    Tstart += count1\n",
    "                elif end1 == n1:\n",
    "                    Tend -= count1\n",
    "                else:\n",
    "                    Tgapcount += 1\n",
    "                    Tgapbases += count1\n",
    "                start1 = end1\n",
    "            else:\n",
    "                assert count1 == count2\n",
    "                tStarts.append(start1)\n",
    "                qStarts.append(start2)\n",
    "                blockSizes.append(count1)\n",
    "                for c1, c2 in zip(seq1[start1:end1], seq2[start2:end2]):\n",
    "                    if c1 == \"N\" or c2 == \"N\":\n",
    "                        Ns += 1\n",
    "                    elif c1 == c2:\n",
    "                        match += 1\n",
    "                    else:\n",
    "                        mismatch += 1\n",
    "                start1 = end1\n",
    "                start2 = end2\n",
    "        blockcount = len(blockSizes)\n",
    "        blockSizes = \",\".join(map(str, blockSizes)) + \",\"\n",
    "        qStarts = \",\".join(map(str, qStarts)) + \",\"\n",
    "        tStarts = \",\".join(map(str, tStarts)) + \",\"\n",
    "        words = [str(match),\n",
    "                 str(mismatch),\n",
    "                 str(repmatch),\n",
    "                 str(Ns),\n",
    "                 str(Qgapcount),\n",
    "                 str(Qgapbases),\n",
    "                 str(Tgapcount),\n",
    "                 str(Tgapbases),\n",
    "                 strand,\n",
    "                 Qname,\n",
    "                 str(Qsize),\n",
    "                 str(Qstart),\n",
    "                 str(Qend),\n",
    "                 Tname,\n",
    "                 str(Tsize),\n",
    "                 str(Tstart),\n",
    "                 str(Tend),\n",
    "                 str(blockcount),\n",
    "                 blockSizes,\n",
    "                 qStarts,\n",
    "                 tStarts,\n",
    "                 ]\n",
    "        line = \"\\t\".join(words) + \"\\n\"\n",
    "        return line\n",
    "\n",
    "    @property\n",
    "    def aligned(self):\n",
    "        \"\"\"Return the indices of subsequences aligned to each other.\n",
    "\n",
    "        This property returns the start and end indices of subsequences\n",
    "        in the target and query sequence that were aligned to each other.\n",
    "        If the alignment between target (t) and query (q) consists of N\n",
    "        chunks, you get two tuples of length N:\n",
    "\n",
    "            (((t_start1, t_end1), (t_start2, t_end2), ..., (t_startN, t_endN)),\n",
    "             ((q_start1, q_end1), (q_start2, q_end2), ..., (q_startN, q_endN)))\n",
    "\n",
    "        For example,\n",
    "\n",
    "        >>> from Bio import Align\n",
    "        >>> aligner = Align.PairwiseAligner()\n",
    "        >>> alignments = aligner.align(\"GAACT\", \"GAT\")\n",
    "        >>> alignment = alignments[0]\n",
    "        >>> print(alignment)\n",
    "        GAACT\n",
    "        ||--|\n",
    "        GA--T\n",
    "        <BLANKLINE>\n",
    "        >>> alignment.aligned\n",
    "        (((0, 2), (4, 5)), ((0, 2), (2, 3)))\n",
    "        >>> alignment = alignments[1]\n",
    "        >>> print(alignment)\n",
    "        GAACT\n",
    "        |-|-|\n",
    "        G-A-T\n",
    "        <BLANKLINE>\n",
    "        >>> alignment.aligned\n",
    "        (((0, 1), (2, 3), (4, 5)), ((0, 1), (1, 2), (2, 3)))\n",
    "\n",
    "        Note that different alignments may have the same subsequences\n",
    "        aligned to each other. In particular, this may occur if alignments\n",
    "        differ from each other in terms of their gap placement only:\n",
    "\n",
    "        >>> aligner.mismatch_score = -10\n",
    "        >>> alignments = aligner.align(\"AAACAAA\", \"AAAGAAA\")\n",
    "        >>> len(alignments)\n",
    "        2\n",
    "        >>> print(alignments[0])\n",
    "        AAAC-AAA\n",
    "        |||--|||\n",
    "        AAA-GAAA\n",
    "        <BLANKLINE>\n",
    "        >>> alignments[0].aligned\n",
    "        (((0, 3), (4, 7)), ((0, 3), (4, 7)))\n",
    "        >>> print(alignments[1])\n",
    "        AAA-CAAA\n",
    "        |||--|||\n",
    "        AAAG-AAA\n",
    "        <BLANKLINE>\n",
    "        >>> alignments[1].aligned\n",
    "        (((0, 3), (4, 7)), ((0, 3), (4, 7)))\n",
    "\n",
    "        The property can be used to identify alignments that are identical\n",
    "        to each other in terms of their aligned sequences.\n",
    "        \"\"\"\n",
    "        segments1 = []\n",
    "        segments2 = []\n",
    "        if sys.version_info[0] > 2:\n",
    "            i1, i2 = self.path[0]\n",
    "            for node in self.path[1:]:\n",
    "                j1, j2 = node\n",
    "                if j1 > i1 and j2 > i2:\n",
    "                    segment1 = (i1, j1)\n",
    "                    segment2 = (i2, j2)\n",
    "                    segments1.append(segment1)\n",
    "                    segments2.append(segment2)\n",
    "                i1, i2 = j1, j2\n",
    "        else:\n",
    "            # Python 2: convert all long ints to ints to be consistent\n",
    "            # with the doctests\n",
    "            i1, i2 = self.path[0]\n",
    "            i1 = int(i1)\n",
    "            i2 = int(i2)\n",
    "            for node in self.path[1:]:\n",
    "                j1, j2 = node\n",
    "                j1 = int(j1)\n",
    "                j2 = int(j2)\n",
    "                if j1 > i1 and j2 > i2:\n",
    "                    segment1 = (i1, j1)\n",
    "                    segment2 = (i2, j2)\n",
    "                    segments1.append(segment1)\n",
    "                    segments2.append(segment2)\n",
    "                i1, i2 = j1, j2\n",
    "        return tuple(segments1), tuple(segments2)\n",
    "\n",
    "\n",
    "class PairwiseAlignments(object):\n",
    "    \"\"\"Implements an iterator over pairwise alignments returned by the aligner.\n",
    "\n",
    "    This class also supports indexing, which is fast for increasing indices,\n",
    "    but may be slow for random access of a large number of alignments.\n",
    "\n",
    "    Note that pairwise aligners can return an astronomical number of alignments,\n",
    "    even for relatively short sequences, if they align poorly to each other. We\n",
    "    therefore recommend to first check the number of alignments, accessible as\n",
    "    len(alignments), which can be calculated quickly even if the number of\n",
    "    alignments is very large.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, seqA, seqB, score, paths):\n",
    "        \"\"\"Initialize a new PairwiseAlignments object.\n",
    "\n",
    "        Arguments:\n",
    "         - seqA  - The first sequence, as a plain string, without gaps.\n",
    "         - seqB  - The second sequence, as a plain string, without gaps.\n",
    "         - score - The alignment score.\n",
    "         - paths - An iterator over the paths in the traceback matrix;\n",
    "                   each path defines one alignment.\n",
    "\n",
    "        You would normally obtain an PairwiseAlignments object by calling\n",
    "        aligner.align(seqA, seqB), where aligner is a PairwiseAligner object.\n",
    "        \"\"\"\n",
    "        self.seqA = seqA\n",
    "        self.seqB = seqB\n",
    "        self.score = score\n",
    "        self.paths = paths\n",
    "        self.index = -1\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.paths)\n",
    "\n",
    "    def __getitem__(self, index):\n",
    "        if index == self.index:\n",
    "            return self.alignment\n",
    "        if index < self.index:\n",
    "            self.paths.reset()\n",
    "            self.index = -1\n",
    "        while self.index < index:\n",
    "            try:\n",
    "                alignment = next(self)\n",
    "            except StopIteration:\n",
    "                raise_from(IndexError(\"index out of range\"), None)\n",
    "        return alignment\n",
    "\n",
    "    def __iter__(self):\n",
    "        self.paths.reset()\n",
    "        self.index = -1\n",
    "        return self\n",
    "\n",
    "    def __next__(self):\n",
    "        path = next(self.paths)\n",
    "        self.index += 1\n",
    "        alignment = PairwiseAlignment(self.seqA, self.seqB, path, self.score)\n",
    "        self.alignment = alignment\n",
    "        return alignment\n",
    "\n",
    "    if sys.version_info[0] < 3:  # Python 2\n",
    "        next = __next__\n",
    "\n",
    "\n",
    "class PairwiseAligner(_aligners.PairwiseAligner):\n",
    "    \"\"\"Performs pairwise sequence alignment using dynamic programming.\n",
    "\n",
    "    This provides functions to get global and local alignments between two\n",
    "    sequences.  A global alignment finds the best concordance between all\n",
    "    characters in two sequences.  A local alignment finds just the\n",
    "    subsequences that align the best.\n",
    "\n",
    "    To perform a pairwise sequence alignment, first create a PairwiseAligner\n",
    "    object.  This object stores the match and mismatch scores, as well as the\n",
    "    gap scores.  Typically, match scores are positive, while mismatch scores\n",
    "    and gap scores are negative or zero.  By default, the match score is 1,\n",
    "    and the mismatch and gap scores are zero.  Based on the values of the gap\n",
    "    scores, a PairwiseAligner object automatically chooses the appropriate\n",
    "    alignment algorithm (the Needleman-Wunsch, Smith-Waterman, Gotoh, or\n",
    "    Waterman-Smith-Beyer global or local alignment algorithm).\n",
    "\n",
    "    Calling the \"score\" method on the aligner with two sequences as arguments\n",
    "    will calculate the alignment score between the two sequences.\n",
    "    Calling the \"align\" method on the aligner with two sequences as arguments\n",
    "    will return a generator yielding the alignments between the two\n",
    "    sequences.\n",
    "\n",
    "    Some examples:\n",
    "\n",
    "    >>> from Bio import Align\n",
    "    >>> aligner = Align.PairwiseAligner()\n",
    "    >>> alignments = aligner.align(\"TACCG\", \"ACG\")\n",
    "    >>> for alignment in sorted(alignments):\n",
    "    ...     print(\"Score = %.1f:\" % alignment.score)\n",
    "    ...     print(alignment)\n",
    "    ...\n",
    "    Score = 3.0:\n",
    "    TACCG\n",
    "    -|-||\n",
    "    -A-CG\n",
    "    <BLANKLINE>\n",
    "    Score = 3.0:\n",
    "    TACCG\n",
    "    -||-|\n",
    "    -AC-G\n",
    "    <BLANKLINE>\n",
    "\n",
    "    Specify the aligner mode as local to generate local alignments:\n",
    "\n",
    "    >>> aligner.mode = 'local'\n",
    "    >>> alignments = aligner.align(\"TACCG\", \"ACG\")\n",
    "    >>> for alignment in sorted(alignments):\n",
    "    ...     print(\"Score = %.1f:\" % alignment.score)\n",
    "    ...     print(alignment)\n",
    "    ...\n",
    "    Score = 3.0:\n",
    "    TACCG\n",
    "     |-||\n",
    "     A-CG\n",
    "    <BLANKLINE>\n",
    "    Score = 3.0:\n",
    "    TACCG\n",
    "     ||-|\n",
    "     AC-G\n",
    "    <BLANKLINE>\n",
    "\n",
    "    Do a global alignment.  Identical characters are given 2 points,\n",
    "    1 point is deducted for each non-identical character.\n",
    "\n",
    "    >>> aligner.mode = 'global'\n",
    "    >>> aligner.match_score = 2\n",
    "    >>> aligner.mismatch_score = -1\n",
    "    >>> for alignment in aligner.align(\"TACCG\", \"ACG\"):\n",
    "    ...     print(\"Score = %.1f:\" % alignment.score)\n",
    "    ...     print(alignment)\n",
    "    ...\n",
    "    Score = 6.0:\n",
    "    TACCG\n",
    "    -||-|\n",
    "    -AC-G\n",
    "    <BLANKLINE>\n",
    "    Score = 6.0:\n",
    "    TACCG\n",
    "    -|-||\n",
    "    -A-CG\n",
    "    <BLANKLINE>\n",
    "\n",
    "    Same as above, except now 0.5 points are deducted when opening a\n",
    "    gap, and 0.1 points are deducted when extending it.\n",
    "\n",
    "    >>> aligner.open_gap_score = -0.5\n",
    "    >>> aligner.extend_gap_score = -0.1\n",
    "    >>> aligner.target_end_gap_score = 0.0\n",
    "    >>> aligner.query_end_gap_score = 0.0\n",
    "    >>> for alignment in aligner.align(\"TACCG\", \"ACG\"):\n",
    "    ...     print(\"Score = %.1f:\" % alignment.score)\n",
    "    ...     print(alignment)\n",
    "    ...\n",
    "    Score = 5.5:\n",
    "    TACCG\n",
    "    -|-||\n",
    "    -A-CG\n",
    "    <BLANKLINE>\n",
    "    Score = 5.5:\n",
    "    TACCG\n",
    "    -||-|\n",
    "    -AC-G\n",
    "    <BLANKLINE>\n",
    "\n",
    "    The alignment function can also use known matrices already included in\n",
    "    Biopython:\n",
    "\n",
    "    >>> from Bio.Align import substitution_matrices\n",
    "    >>> aligner = Align.PairwiseAligner()\n",
    "    >>> aligner.substitution_matrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "    >>> alignments = aligner.align(\"KEVLA\", \"EVL\")\n",
    "    >>> alignments = list(alignments)\n",
    "    >>> print(\"Number of alignments: %d\" % len(alignments))\n",
    "    Number of alignments: 1\n",
    "    >>> alignment = alignments[0]\n",
    "    >>> print(\"Score = %.1f\" % alignment.score)\n",
    "    Score = 13.0\n",
    "    >>> print(alignment)\n",
    "    KEVLA\n",
    "    -|||-\n",
    "    -EVL-\n",
    "    <BLANKLINE>\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __setattr__(self, key, value):\n",
    "        if key not in dir(_aligners.PairwiseAligner):\n",
    "            # To prevent confusion, don't allow users to create new attributes\n",
    "            message = \"'PairwiseAligner' object has no attribute '%s'\" % key\n",
    "            raise AttributeError(message)\n",
    "        _aligners.PairwiseAligner.__setattr__(self, key, value)\n",
    "\n",
    "    def align(self, seqA, seqB):\n",
    "        \"\"\"Return the alignments of two sequences using PairwiseAligner.\"\"\"\n",
    "        if isinstance(seqA, Seq):\n",
    "            seqA = str(seqA)\n",
    "        if isinstance(seqB, Seq):\n",
    "            seqB = str(seqB)\n",
    "        score, paths = _aligners.PairwiseAligner.align(self, seqA, seqB)\n",
    "        alignments = PairwiseAlignments(seqA, seqB, score, paths)\n",
    "        return alignments\n",
    "\n",
    "    def score(self, seqA, seqB):\n",
    "        \"\"\"Return the alignments score of two sequences using PairwiseAligner.\"\"\"\n",
    "        if isinstance(seqA, Seq):\n",
    "            seqA = str(seqA)\n",
    "        if isinstance(seqB, Seq):\n",
    "            seqB = str(seqB)\n",
    "        return _aligners.PairwiseAligner.score(self, seqA, seqB)\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    from Bio._utils import run_doctest\n",
    "    run_doctest()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
